<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content="summary" name="twitter:card" />
<meta content="&#64;immuneml" name="twitter:site" />
<meta content="immuneML use case: extending immuneML with a deep learning component for predicting antigen specificity of paired receptor data" name="twitter:title" />
<meta content="See how immuneML was extended with a deep learning component for predicting antigen specificity of paired receptor data in the immuneML manuscript." name="twitter:description" />
<meta content="https://docs.immuneml.uio.no/_images/epitope_1.png" name="twitter:image" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Manuscript use case 3: Benchmarking ML methods on ground-truth synthetic data" href="benchmarking_use_case.html" /><link rel="prev" title="Manuscript use case 1: Reproduction of a published study inside immuneML" href="emerson_reproduction.html" />
        <link rel="canonical" href="https://docs.immuneml.uio.no/usecases/extendability_use_case.html" />

    <link rel="shortcut icon" href="../_static/favicon.ico"/><!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>Manuscript use case 2: Extending immuneML with a deep learning component for predicting antigen specificity of paired receptor data - immuneML 3.0.14 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/immuneml.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">immuneML 3.0.14 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">immuneML 3.0.14 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 has-children"><a class="reference internal" href="../quickstart.html">Quickstart</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Quickstart</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/galaxy_simple.html">Quickstart: Galaxy with button-based tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/galaxy_yaml.html">Quickstart: Galaxy with YAML-based tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/cli_yaml.html">Quickstart: command-line interface with YAML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/simulation_quickstart.html">LIgO simulation quickstart</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../installation.html">Installing immuneML</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Installing immuneML</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation/install_with_package_manager.html">Install immuneML with a package manager</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/installation_docker.html">Setting up immuneML with Docker</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/cloud.html">Running immuneML in the cloud</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../specification.html">YAML specification</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of YAML specification</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/how_to_specify_an_analysis_with_yaml.html">How to specify an analysis with YAML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/datasets.html">Dataset parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/encodings.html">Encoding parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/ml_methods.html">ML method parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/reports.html">Report parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/preprocessings.html">Preprocessing parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/simulation.html">Simulation parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/instructions.html">Instruction parameters</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/analyze_dataset.html">Analyzing Your Own Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_import_the_data_to_immuneML.html">How to import data into immuneML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_generate_a_random_repertoire_dataset.html">How to generate a dataset with random sequences</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../tutorials/ligo_simulation_tutorials.html">Dataset simulation with LIgO</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of Dataset simulation with LIgO</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/ligo_simulation_yaml.html">YAML specification of the LigoSim instruction for introducing immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/how_to_simulate_co-occuring_signals.html">How to simulate co-occuring immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/how_to_simulate_paired_chain_data.html">Paired chain simulations in LIgO</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/simulation_with_custom_signal_functions.html">Simulation with custom signal functions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_train_and_assess_a_receptor_or_repertoire_classifier.html">How to train and assess a receptor or repertoire-level ML classifier</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_apply_to_new_data.html">How to apply previously trained ML models to a new dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_perform_exploratory_analysis.html">How to perform an exploratory data analysis</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../tutorials/motif_recovery.html">How to find motifs associated with disease or antigen binding state</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of How to find motifs associated with disease or antigen binding state</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/discover_motifs_precision_recall.html">Discovering positional motifs using precision and recall thresholds</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/discover_motifs_classifiers.html">Discovering motifs learned by classifiers</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/recovering_simulated_motifs.html">Recovering simulated immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/comparing_baseline_motifs.html">Comparing baseline motif frequencies in repertoires</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/clustering_tutorial.html">How to perform clustering analysis</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../galaxy.html">immuneML &amp; Galaxy</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of immuneML &amp; Galaxy</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_intro.html">Introduction to Galaxy</a></li>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_tools.html">immuneML Galaxy tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_immunologist_friendly.html">ML basics: Training classifiers with the simplified Galaxy interface</a></li>
</ul>
</li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../usecases.html">Use case examples</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of Use case examples</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="emerson_reproduction.html">Manuscript use case 1: Reproduction of a published study inside immuneML</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Manuscript use case 2: Extending immuneML with a deep learning component for predicting antigen specificity of paired receptor data</a></li>
<li class="toctree-l2"><a class="reference internal" href="benchmarking_use_case.html">Manuscript use case 3: Benchmarking ML methods on ground-truth synthetic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="immcantation_use_case.html">Integration use case: post-analysis of sequences with Immcantation</a></li>
<li class="toctree-l2"><a class="reference internal" href="immunarch_use_case.html">Integration use case: post-analysis of sequences with immunarch</a></li>
<li class="toctree-l2"><a class="reference internal" href="immunesim_use_case.html">Integration use case: Performing analysis on immuneSIM-generated repertoires</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../troubleshooting.html">Troubleshooting</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of Troubleshooting</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../developer_docs.html">Developer documentation</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of Developer documentation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/info_new_developers.html">Information for new developers</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/install_for_development.html">Set up immuneML for development</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_encoding.html">How to add a new encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_ML_method.html">How to add a new machine learning method</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_report.html">How to add a new report</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_preprocessing.html">How to add a new preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/data_model.html">immuneML data model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/execution_flow.html">immuneML execution flow</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/usecases/extendability_use_case.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="manuscript-use-case-2-extending-immuneml-with-a-deep-learning-component-for-predicting-antigen-specificity-of-paired-receptor-data">
<h1>Manuscript use case 2: Extending immuneML with a deep learning component for predicting antigen specificity of paired receptor data<a class="headerlink" href="#manuscript-use-case-2-extending-immuneml-with-a-deep-learning-component-for-predicting-antigen-specificity-of-paired-receptor-data" title="Link to this heading">¶</a></h1>
<p>To demonstrate how immuneML can be extended, we added a convolutional neural network (CNN) for predicting receptor specificity. To show how it can be
using alongside existing components, we compared the CNN with k-nearest neighbour classifier based on TCRdist distance and a
logistic regression model on 3-mer frequencies. We also show the motifs extracted by CNN, TCRdist
(<a class="reference external" href="https://www.biorxiv.org/content/10.1101/2020.12.24.424260v1">Mayer-Blackwell et al. 2020</a>, <a class="reference external" href="https://www.nature.com/articles/nature22383">Dash et al. 2017</a>) and GLIPH2
(<a class="reference external" href="https://www.nature.com/articles/s41587-020-0505-4">Huang et al. 2020</a>) for the datasets we examine.</p>
<p>This use-case description includes the description of the datasets, how to add a CNN model, how to compare CNN with other models and finally, how to
extract motifs from models.</p>
<p>The complete collection of original files used in this use case can be found in the NIRD research data archive (DOI: <a class="reference external" href="https://doi.org/10.11582/2021.00009">10.11582/2021.00009</a>).
Note that the YAML specifications in the original dataset were compatible with immuneML version 1.1.1.
This documentation page contains the YAML specifications for equivalent analyses with the latest immuneML version (last tested with version 1.1.4).</p>
<section id="preparing-the-datasets">
<h2>Preparing the datasets<a class="headerlink" href="#preparing-the-datasets" title="Link to this heading">¶</a></h2>
<p>We used three epitope-specific datasets to compare the performance of CNN, TCRdist classifier and logistic regression. The epitope-specific paired chain
receptors were downloaded from VDJdb for epitopes:</p>
<ul class="simple">
<li><p>AVFDRKSDAK (Epstein-Barr virus): the original file is available here: <code class="xref download docutils literal notranslate"><span class="pre">AVFDRKSDAK.tsv</span></code></p></li>
<li><p>GILGFVFTL (Influenza A): the original file is available here: <code class="xref download docutils literal notranslate"><span class="pre">GILGFVFTL.tsv</span></code></p></li>
<li><p>KLGGALQAK (Cytomegalovirus): the original file is available here: <code class="xref download docutils literal notranslate"><span class="pre">KLGGALQAK.tsv</span></code></p></li>
</ul>
<p>Naive unpaired T-cell receptor data was downloaded from the immuneACCESS database from the study by <a class="reference external" href="http://www.sciencedirect.com/science/article/pii/S016158902030479X">Heikkilä et al. 2020</a>. It consisted of naive PBMC
TCRAB data from 4 healthy donors that where then randomly paired if coming from the same donor, as described also by Dash et al. 2017 for generation of
background receptors. The receptors missing the V- or J-gene information (even though in some cases gene subgroup might have been know),
receptors shorter than 6 amino acids and receptors containing V- or J-gene information for which the mapping was not available in TCRdist3 were
removed. The removed genes due to TCRdist3 availablility were: TCRAV37-01*01, TCRBV07-05*02, TCRAV28-01, TCRAV31-01*01, TCRBV07-05, TCRAV33-01*01,
TCRAV32-01*01, TCRBVA-or09_02, TCRBV22-01.</p>
<p>The original data file as downloaded from immuneACCESS is available here: <code class="xref download docutils literal notranslate"><span class="pre">naiveTCRAB.tsv</span></code></p>
<p>Using this data, three datasets were created by combining epitope-specific paired receptor data with naive randomly paired data. These data along with
the results of testing are available in <a class="reference external" href="http://doi.org/10.11582/2021.00009">NIRD research data archive</a>.</p>
</section>
<section id="adding-a-new-ml-method">
<h2>Adding a new ML method<a class="headerlink" href="#adding-a-new-ml-method" title="Link to this heading">¶</a></h2>
<p>We proposed a CNN model adapted to the receptor specificity prediction problem. Briefly, an adaptive immune receptor consists of two chains, two receptor
(amino acid) sequences forming a 3D structure. This receptor can either bind an antigen (e.g., part of a virus) and trigger an immune response or not.
The ability to predict if receptor binds an antigen would provide useful knowledge for diagnostics and therapeutics development.</p>
<p>The CNN we proposed for this task would consist of one set of kernels of multiple kernel sizes for each receptor chain. The kernels are followed by
ReLU activation and max-pooling to obtain one representation vector per receptor chain. These two vectors are then concatenated and form a unique
receptor representation. This receptor representation is then provided as input for a single linear layer with a single output predicting if a
receptor will bind an antigen or not.</p>
<p>To add a new ML method, we added two classes to the <code class="code docutils literal notranslate"><span class="pre">ml_methods</span></code> package:</p>
<ol class="arabic simple">
<li><p>ReceptorCNN class implementing MLMethod interface and</p></li>
<li><p>PyTorchReceptorCNN class inheriting PyTorch’s nn.Module and implementing the forward pass through the network.</p></li>
</ol>
<p>The <code class="code docutils literal notranslate"><span class="pre">ReceptorCNN</span></code> class is the main class given in the YAML specification when this method is used, while the <code class="code docutils literal notranslate"><span class="pre">PyTorchReceptorCNN</span></code> class is PyTorch-specific
implementation. <code class="code docutils literal notranslate"><span class="pre">ReceptorCNN</span></code> creates an object of the <code class="code docutils literal notranslate"><span class="pre">PyTorchReceptorCNN</span></code> class and takes care of training and predicting class assignment for
training instances. See the documentation and the code for ReceptorCNN at <code class="xref py py-obj docutils literal notranslate"><span class="pre">immuneML.ml_methods.ReceptorCNN.ReceptorCNN</span></code> and for PyTorchReceptorCNN at
<code class="xref py py-obj docutils literal notranslate"><span class="pre">immuneML.ml_methods.pytorch_implementations.PyTorchReceptorCNN.PyTorchReceptorCNN</span></code>. We also provide a tutorial <a class="reference internal" href="../developer_docs/how_to_add_new_ML_method.html#how-to-add-a-new-machine-learning-method"><span class="std std-ref">How to add a new machine learning method</span></a>.</p>
</section>
<section id="comparing-the-ml-models-and-recovered-motifs-from-tcrdist3-gliph2-and-cnn">
<h2>Comparing the ML models and recovered motifs from TCRdist3, GLIPH2 and CNN<a class="headerlink" href="#comparing-the-ml-models-and-recovered-motifs-from-tcrdist3-gliph2-and-cnn" title="Link to this heading">¶</a></h2>
<p>We compared three ML methods to predict if a paired receptor will bind an antigen or not:</p>
<ol class="arabic simple">
<li><p>ReceptorCNN method added in the previous step,</p></li>
<li><p>k-nearest neighbors classifier based on TCRdist distance between paired receptors, and</p></li>
<li><p>logistic regression model based on 3-mer frequencies in each chain.</p></li>
</ol>
<p>We compared the performance on three datasets described previously. The YAML specification for one of these analysis (for Influenza A, epitope GILGFVFTL)
is available below. In addition to training and comparing the models, we also recovered motifs from each of them to show how we could get some
insight into the inner workings of each method. We used immuneML to extract the motifs from the trained TCRdist-based classifier, logistic regression
model and CNN. Additionally, we exported the data in a format compatible with GLIPH2 from immuneML and used the same data to extract the motifs with
GLIPH2 as for the other models.</p>
<p>To recover the motifs from the ReceptorCNN model, we added a <a class="reference internal" href="../yaml_specs/reports.html#kernelsequencelogo"><span class="std std-ref">KernelSequenceLogo</span></a> report to visualize the kernels of the network as information
gain matrices, similar to the work of <a class="reference external" href="https://www.biorxiv.org/content/10.1101/411934v1">Ploenzke and Irizarry 2018</a>. Adding a report to explore machine learning
models is described in the tutorial <a class="reference internal" href="../developer_docs/how_to_add_new_report.html#how-to-add-a-new-report"><span class="std std-ref">How to add a new report</span></a>. Briefly, adding a report to explore the motifs in the CNN was done by implementing and adding
a class that implements the given report, without modifying any other part of the code.</p>
<p>The input data, YAML specifications and results of these comparisons are available in <a class="reference external" href="http://doi.org/10.11582/2021.00009">NIRD research data archive</a>.</p>
<p>YAML specification for training machine learning models and recovering motifs in the GILGFVFTL dataset (note that the ROCCurveSummary was not present in the original
specification and the plots were obtained outside immuneML; it has since been added to immuneML and can be used directly from YAML specification):</p>
<blockquote>
<div><details class="summary-model-comparison-yaml">
<summary>model_comparison.yaml</summary><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">definitions</span><span class="p">:</span><span class="w"> </span><span class="c1"># definition of the components: datasets, encodings, ML models, reports</span>
<span class="w">  </span><span class="nt">datasets</span><span class="p">:</span>
<span class="w">    </span><span class="nt">dataset_GILGFVFTL</span><span class="p">:</span><span class="w"> </span><span class="c1"># the dataset specific for the epitope GILGFVFTL of EBV virus combined with naive receptors</span>
<span class="w">      </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AIRR</span>
<span class="w">      </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">is_repertoire</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="w">        </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">data/GILGFVFTL.tsv</span>
<span class="w">        </span><span class="nt">paired</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">        </span><span class="nt">import_illegal_characters</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="w">        </span><span class="nt">import_empty_nt_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">        </span><span class="nt">import_empty_aa_sequences</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="w">        </span><span class="nt">receptor_chains</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TRA_TRB</span>
<span class="w">        </span><span class="nt">metadata_column_mapping</span><span class="p">:</span>
<span class="w">          </span><span class="nt">GILGFVFTL</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GILGFVFTL</span>
<span class="w">        </span><span class="nt">organism</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">human</span>
<span class="w">  </span><span class="nt">encodings</span><span class="p">:</span><span class="w"> </span><span class="c1"># how to represent the data</span>
<span class="w">    </span><span class="nt">one_hot</span><span class="p">:</span><span class="w"> </span><span class="c1"># encoding for the neural network includes one-hot encoding of receptors and additional three positional channels</span>
<span class="w">      </span><span class="nt">OneHot</span><span class="p">:</span>
<span class="w">        </span><span class="nt">use_positional_info</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>
<span class="w">        </span><span class="nt">distance_to_seq_middle</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span>
<span class="w">        </span><span class="nt">flatten</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
<span class="w">    </span><span class="nt">tcrdist_enc</span><span class="p">:</span><span class="w"> </span><span class="c1"># represents the the receptors by tcrdist distances between them</span>
<span class="w">      </span><span class="nt">TCRdist</span><span class="p">:</span>
<span class="w">        </span><span class="nt">cores</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">32</span>
<span class="w">    </span><span class="nt">kmer_frequency</span><span class="p">:</span><span class="w"> </span><span class="c1"># represents receptors by the frequency of k-mers per chain</span>
<span class="w">      </span><span class="nt">KmerFrequency</span><span class="p">:</span>
<span class="w">        </span><span class="nt">k</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span>
<span class="w">        </span><span class="nt">sequence_encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">continuous_kmer</span><span class="w">  </span><span class="c1"># split sequence into overlapping k-mers</span>
<span class="w">        </span><span class="nt">scale_to_unit_variance</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># scale the normalized examples to have unit variance</span>
<span class="w">        </span><span class="nt">scale_to_zero_mean</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># scale the normalized examples to have zero mean -&gt; setting this to True might destroy sparsity</span>
<span class="w">  </span><span class="nt">ml_methods</span><span class="p">:</span><span class="w"> </span><span class="c1"># which machine learning methods to use</span>
<span class="w">    </span><span class="nt">cnn</span><span class="p">:</span><span class="w"> </span><span class="c1"># a convolutiional network</span>
<span class="w">      </span><span class="nt">ReceptorCNN</span><span class="p">:</span>
<span class="w">        </span><span class="nt">number_of_threads</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">32</span>
<span class="w">        </span><span class="nt">batch_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20000</span>
<span class="w">        </span><span class="nt">evaluate_at</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>
<span class="w">        </span><span class="nt">iteration_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20000</span>
<span class="w">        </span><span class="nt">kernel_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">50</span>
<span class="w">        </span><span class="nt">kernel_size</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">3</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">4</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">5</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">6</span><span class="p p-Indicator">]</span>
<span class="w">        </span><span class="nt">positional_channels</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span>
<span class="w">        </span><span class="nt">learning_rate</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.01</span>
<span class="w">        </span><span class="nt">l1_weight_decay</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.01</span>
<span class="w">        </span><span class="nt">l2_weight_decay</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.01</span>
<span class="w">    </span><span class="nt">tcrdist_cls</span><span class="p">:</span><span class="w"> </span><span class="c1"># kNN classifier based on tcrdist distances [computed in the encoding phase]</span>
<span class="w">      </span><span class="nt">TCRdistClassifier</span><span class="p">:</span>
<span class="w">        </span><span class="nt">percentage</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.1</span><span class="w"> </span><span class="c1"># the percentage of receptors relevant for determining the class assignment (binding vs. not binding), based on the paper by Dash $</span>
<span class="w">    </span><span class="nt">logistic_regression</span><span class="p">:</span>
<span class="w">      </span><span class="nt">LogisticRegression</span><span class="p">:</span>
<span class="w">        </span><span class="nt">penalty</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">l1</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">l2</span><span class="p p-Indicator">]</span><span class="w"> </span><span class="c1"># try lasso and ridge</span>
<span class="w">        </span><span class="nt">C</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">1000</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">100</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">10</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">1</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">0.1</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">0.01</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">0.001</span><span class="p p-Indicator">]</span><span class="w"> </span><span class="c1"># regularization constants</span>
<span class="w">      </span><span class="nt">model_selection_cv</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># do a third level of cross-validation where only the logistic regression hyperparameters listed here will be optimized by $</span>
<span class="w">      </span><span class="nt">model_selection_n_folds</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span><span class="w"> </span><span class="c1"># for this third level of CV, use 5-fold CV</span>
<span class="w">  </span><span class="nt">reports</span><span class="p">:</span><span class="w"> </span><span class="c1"># additional output: analysis reports</span>
<span class="w">    </span><span class="nt">cnn_kernel_logo</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">KernelSequenceLogo</span><span class="w"> </span><span class="c1"># this report will visualize kernels from the CNN as sequences logos since the kernels are conditioned to represen$</span>
<span class="w">    </span><span class="nt">tcrdist_logo</span><span class="p">:</span><span class="w"> </span><span class="c1"># uses tcrdist3 library to discover the motifs in the hierarchically clustered receptors based on tcrdist distance</span>
<span class="w">      </span><span class="nt">TCRdistMotifDiscovery</span><span class="p">:</span>
<span class="w">        </span><span class="nt">positive_class_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w"> </span><span class="c1"># will only cluster the positive class (receptors binding to GILGFVFTL)</span>
<span class="w">        </span><span class="nt">min_cluster_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span><span class="w"> </span><span class="c1"># minimum cluster size to extract the motif from</span>
<span class="w">    </span><span class="nt">coefficients</span><span class="p">:</span>
<span class="w">      </span><span class="nt">Coefficients</span><span class="p">:</span><span class="w"> </span><span class="c1"># plot top 100 largest coefficients for logistic regression</span>
<span class="w">        </span><span class="nt">coefs_to_plot</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">n_largest</span><span class="p p-Indicator">]</span>
<span class="w">        </span><span class="nt">n_largest</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">100</span><span class="p p-Indicator">]</span>
<span class="w">    </span><span class="nt">gliph_exporter</span><span class="p">:</span><span class="w"> </span><span class="c1"># will export the receptor data in format compatible with GLIPH2 (Huang et al. 2020) so that it can be directly used on the data as sp$</span>
<span class="w">      </span><span class="nt">GLIPH2Exporter</span><span class="p">:</span>
<span class="w">        </span><span class="nt">condition</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GILGFVFTL</span><span class="w"> </span><span class="c1"># what is the condition, as defined by GLIPH2</span>
<span class="w">    </span><span class="nt">roc_summary</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ROCCurveSummary</span><span class="w"> </span><span class="c1"># plot the ROC curves on the performance for each split in the outer cross-validation loop to compare different methods defined here</span>
<span class="nt">instructions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">tcrdist_cnn_comparison</span><span class="p">:</span><span class="w"> </span><span class="c1"># definition of the analysis</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TrainMLModel</span><span class="w"> </span><span class="c1"># analysis for training ML models</span>
<span class="w">    </span><span class="nt">settings</span><span class="p">:</span><span class="w"> </span><span class="c1"># which combinations of previously defined encodings and ML models to consider</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">one_hot</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cnn</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">tcrdist_enc</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">tcrdist_cls</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">kmer_frequency</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">logistic_regression</span>
<span class="w">    </span><span class="nt">assessment</span><span class="p">:</span><span class="w"> </span><span class="c1"># how to perform the outer loop of nested cross-validation to obtain performance estimate</span>
<span class="w">      </span><span class="nt">split_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">random</span><span class="w"> </span><span class="c1"># randomly split to train and test</span>
<span class="w">      </span><span class="nt">split_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"> </span><span class="c1"># make only one such split</span>
<span class="w">      </span><span class="nt">training_percentage</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span><span class="w"> </span><span class="c1"># use 70% of data for testing</span>
<span class="w">      </span><span class="nt">reports</span><span class="p">:</span><span class="w"> </span><span class="c1"># which reports to generate of the data and the models obtained in this way</span>
<span class="w">        </span><span class="nt">models</span><span class="p">:</span><span class="w"> </span><span class="c1"># reports to generate from the models, here: motif discovery</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cnn_kernel_logo</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">tcrdist_logo</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">coefficients</span>
<span class="w">        </span><span class="nt">data_splits</span><span class="p">:</span><span class="w"> </span><span class="c1"># reports to generate from the data, here: output the data in GLIPH2-compatible format</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">gliph_exporter</span>
<span class="w">    </span><span class="nt">selection</span><span class="p">:</span><span class="w"> </span><span class="c1"># how to perform the inner loop of nested cross-validation to choose the optimal model</span>
<span class="w">      </span><span class="nt">split_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">k_fold</span><span class="w"> </span><span class="c1"># do 5-fold cross-validation</span>
<span class="w">      </span><span class="nt">split_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">      </span><span class="nt">reports</span><span class="p">:</span><span class="w"> </span><span class="c1"># again, which reports to generate, same as under assessment</span>
<span class="w">        </span><span class="nt">models</span><span class="p">:</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cnn_kernel_logo</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">tcrdist_logo</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">coefficients</span>
<span class="w">        </span><span class="nt">data_splits</span><span class="p">:</span>
<span class="w">          </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">gliph_exporter</span>
<span class="w">    </span><span class="nt">labels</span><span class="p">:</span><span class="w"> </span><span class="c1"># which labels to use to train the ML models</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GILGFVFTL</span>
<span class="w">    </span><span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset_GILGFVFTL</span><span class="w"> </span><span class="c1"># which dataset to use</span>
<span class="w">    </span><span class="nt">metrics</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">balanced_accuracy</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">precision</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">recall</span><span class="p p-Indicator">]</span><span class="w"> </span><span class="c1"># metrics to be computed for all settings</span>
<span class="w">    </span><span class="nt">strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GridSearch</span><span class="w"> </span><span class="c1"># how to evaluate different combinations of encodings and ML models listed under settings, here: just compare them all with each other</span>
<span class="w">    </span><span class="nt">number_of_processes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">32</span><span class="w"> </span><span class="c1"># in the parallelized parts of the code, how many processes to use</span>
<span class="w">    </span><span class="nt">optimization_metric</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">auc</span><span class="w"> </span><span class="c1"># the metric used for optimization</span>
<span class="w">    </span><span class="nt">reports</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">roc_summary</span><span class="p p-Indicator">]</span><span class="w"> </span><span class="c1"># additional reports made on the full results -&gt; here produce the ROC curves for each method as shown in the results</span>
<span class="w">    </span><span class="nt">refit_optimal_model</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># whether to refit the optimal model before exporting it (not in this use-case as the models will be used for comparison, not for classifying some new data)</span>
</pre></div>
</div>
</details></div></blockquote>
<section id="gliph2-motifs">
<h3>GLIPH2 motifs<a class="headerlink" href="#gliph2-motifs" title="Link to this heading">¶</a></h3>
<p>Running the previous YAML specification will also output the data in GLIPH2-compatible format. To obtain motifs, we filtered the training data
exported in this way for positive examples and provided them as input to GLIPH2 web server available at
<a class="reference external" href="http://50.255.35.37:8080/">the project website</a>. To obtain the motif file, we used GLIPH2 version of the algorithm, reference version 2, CD8
reference, and assumed all amino acids are interchangeable. The file provided as input to GLIPH2, as well as the resulting file, are available in the
<a class="reference external" href="http://doi.org/10.11582/2021.00009">NIRD research data archive</a>.</p>
</section>
</section>
<section id="results">
<h2>Results<a class="headerlink" href="#results" title="Link to this heading">¶</a></h2>
<p>We compared the CNN method with the TCRdist-based k-nearest neighbor classifier and logistic regression on a dataset
consisting of epitope-specific and naive TCRαβ sequences (assumed to be non-epitope-specific).</p>
<p>The resulting ROC curves for the three epitopes (GILGFVFTL, Influenza A; AVFDRKSDAK, Epstein-Barr virus; KLGGALQAK, Cytomegalovirus) are shown below.
Note that these figures were not included in the results deposited in the NIRD research data archive, but the YAML specification above
has been updated to produce these plots.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="../_images/epitope_1.png"><img alt="Epitope 1" src="../_images/epitope_1.png" style="width: 60%;" />
</a>
</figure>
<figure class="align-default">
<a class="reference internal image-reference" href="../_images/epitope_2.png"><img alt="Epitope 2" src="../_images/epitope_2.png" style="width: 60%;" />
</a>
</figure>
<figure class="align-default">
<a class="reference internal image-reference" href="../_images/epitope_3.png"><img alt="Epitope 3" src="../_images/epitope_3.png" style="width: 60%;" />
</a>
</figure>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="benchmarking_use_case.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Manuscript use case 3: Benchmarking ML methods on ground-truth synthetic data</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="emerson_reproduction.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Manuscript use case 1: Reproduction of a published study inside immuneML</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2025, Milena Pavlovic, Lonneke Scheffer, Keshav Motwani, Victor Greiff, Geir Kjetil Sandve
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Manuscript use case 2: Extending immuneML with a deep learning component for predicting antigen specificity of paired receptor data</a><ul>
<li><a class="reference internal" href="#preparing-the-datasets">Preparing the datasets</a></li>
<li><a class="reference internal" href="#adding-a-new-ml-method">Adding a new ML method</a></li>
<li><a class="reference internal" href="#comparing-the-ml-models-and-recovered-motifs-from-tcrdist3-gliph2-and-cnn">Comparing the ML models and recovered motifs from TCRdist3, GLIPH2 and CNN</a><ul>
<li><a class="reference internal" href="#gliph2-motifs">GLIPH2 motifs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#results">Results</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=2efb4362"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>