<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="How to import data into immuneML" href="how_to_import_the_data_to_immuneML.html" /><link rel="prev" title="Tutorials" href="../tutorials.html" />
        <link rel="canonical" href="https://docs.immuneml.uio.no/tutorials/analyze_dataset.html" />

    <link rel="shortcut icon" href="../_static/favicon.ico"/><!-- Generated with Sphinx 8.1.3 and Furo 2024.08.06 -->
        <title>Analyzing Your Own Dataset - immuneML 3.0.18 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/immuneml.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">immuneML 3.0.18 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">immuneML 3.0.18 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 has-children"><a class="reference internal" href="../quickstart.html">Quickstart</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Quickstart</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/galaxy_simple.html">Quickstart: Galaxy with button-based tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/galaxy_yaml.html">Quickstart: Galaxy with YAML-based tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/cli_yaml.html">Quickstart: command-line interface with YAML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/simulation_quickstart.html">LIgO simulation quickstart</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../installation.html">Installing immuneML</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Installing immuneML</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation/install_with_package_manager.html">Install immuneML with a package manager</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/installation_docker.html">Setting up immuneML with Docker</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/cloud.html">Running immuneML in the cloud</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../specification.html">YAML specification</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of YAML specification</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/how_to_specify_an_analysis_with_yaml.html">How to specify an analysis with YAML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/datasets.html">Dataset parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/encodings.html">Encoding parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/ml_methods.html">ML method parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/reports.html">Report parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/preprocessings.html">Preprocessing parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/simulation.html">Simulation parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/instructions.html">Instruction parameters</a></li>
</ul>
</li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Analyzing Your Own Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_import_the_data_to_immuneML.html">How to import data into immuneML</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_generate_a_random_repertoire_dataset.html">How to generate a dataset with random sequences</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="ligo_simulation_tutorials.html">Dataset simulation with LIgO</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of Dataset simulation with LIgO</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="ligo_simulation_yaml.html">YAML specification of the LigoSim instruction for introducing immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="how_to_simulate_co-occuring_signals.html">How to simulate co-occuring immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="how_to_simulate_paired_chain_data.html">Paired chain simulations in LIgO</a></li>
<li class="toctree-l3"><a class="reference internal" href="simulation_with_custom_signal_functions.html">Simulation with custom signal functions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="how_to_train_and_assess_a_receptor_or_repertoire_classifier.html">How to train and assess a receptor or repertoire-level ML classifier</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_apply_to_new_data.html">How to apply previously trained ML models to a new dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_perform_exploratory_analysis.html">How to perform an exploratory data analysis</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="motif_recovery.html">How to find motifs associated with disease or antigen binding state</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of How to find motifs associated with disease or antigen binding state</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="discover_motifs_precision_recall.html">Discovering positional motifs using precision and recall thresholds</a></li>
<li class="toctree-l3"><a class="reference internal" href="discover_motifs_classifiers.html">Discovering motifs learned by classifiers</a></li>
<li class="toctree-l3"><a class="reference internal" href="recovering_simulated_motifs.html">Recovering simulated immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="comparing_baseline_motifs.html">Comparing baseline motif frequencies in repertoires</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="clustering_tutorial.html">How to perform clustering analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_train_and_apply_gen_model.html">How to train a generative model</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_combine_multiple_encodings.html">How to combine multiple encodings to represent a dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_publication_ready_figures.html">How to get publication-ready figures</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../galaxy.html">immuneML &amp; Galaxy</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of immuneML &amp; Galaxy</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_intro.html">Introduction to Galaxy</a></li>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_tools.html">immuneML Galaxy tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_immunologist_friendly.html">ML basics: Training classifiers with the simplified Galaxy interface</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../usecases.html">Use case examples</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of Use case examples</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../usecases/emerson_reproduction.html">Manuscript use case 1: Reproduction of a published study inside immuneML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/extendability_use_case.html">Manuscript use case 2: Extending immuneML with a deep learning component for predicting antigen specificity of paired receptor data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/benchmarking_use_case.html">Manuscript use case 3: Benchmarking ML methods on ground-truth synthetic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immcantation_use_case.html">Integration use case: post-analysis of sequences with Immcantation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immunarch_use_case.html">Integration use case: post-analysis of sequences with immunarch</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immunesim_use_case.html">Integration use case: Performing analysis on immuneSIM-generated repertoires</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../troubleshooting.html">Troubleshooting</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of Troubleshooting</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../developer_docs.html">Developer documentation</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of Developer documentation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/info_new_developers.html">Information for new developers</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/install_for_development.html">Set up immuneML for development</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_encoding.html">How to add a new encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_ML_method.html">How to add a new machine learning method</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_report.html">How to add a new report</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_preprocessing.html">How to add a new preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/data_model.html">immuneML data model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/execution_flow.html">immuneML execution flow</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/tutorials/analyze_dataset.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="analyzing-your-own-dataset">
<h1>Analyzing Your Own Dataset<a class="headerlink" href="#analyzing-your-own-dataset" title="Link to this heading">¶</a></h1>
<p>In this tutorial, we will show how to analyze your own dataset with some useful guidelines on which concrete steps to take,
which instructions to use, what visualizations to create. This tutorial assumes immuneML is already installed at your machine.
If that is not the case, see <a class="reference internal" href="../installation.html#installing-immuneml"><span class="std std-ref">Installing immuneML</span></a>.</p>
<p>The examples in this tutorial will use a <a class="reference download internal" download="" href="../_downloads/6911aa566b35dbe7ea0b8690bcb6ac70/dataset.zip"><code class="xref download docutils literal notranslate"><span class="pre">simulated</span> <span class="pre">dataset</span></code></a>. It consists of 200 repertoires
with two labels: disease (True, False) and batch (True, False).</p>
<p>The available analysis specifications should be general enough that it can be applied to any dataset with
minimal changes (e.g., paths, label names).</p>
<section id="step-1-explore-the-dataset">
<h2>Step 1: Explore the dataset<a class="headerlink" href="#step-1-explore-the-dataset" title="Link to this heading">¶</a></h2>
<p>Before doing any machine learning, it is important to check the dataset. For that purpose, immuneML offers some
useful reports and visualizations:</p>
<ul class="simple">
<li><p><a class="reference internal" href="../yaml_specs/reports.html#aminoacidfrequencydistribution"><span class="std std-ref">AminoAcidFrequencyDistribution</span></a>: shows the relative frequency of each amino acid at each position in the sequences;
the report can also split the dataset per some predefined label (e.g., by disease status)</p></li>
<li><p><a class="reference internal" href="../yaml_specs/reports.html#labeloverlap"><span class="std std-ref">LabelOverlap</span></a>: shows the overlap of two different labels in the dataset as a heatmap; useful if in addition to the
labels you are interested in, you have some other labels (e.g., batch information) that you want to check for overlap</p></li>
<li><p><a class="reference internal" href="../yaml_specs/reports.html#sequencelengthdistribution"><span class="std std-ref">SequenceLengthDistribution</span></a>: shows the histogram of the lengths of sequences by the dataset, optionally split by a
label of interest; useful to see if the dataset is balanced in terms of sequence lengths</p></li>
<li><p><a class="reference internal" href="../yaml_specs/reports.html#vjgenedistribution"><span class="std std-ref">VJGeneDistribution</span></a>: shows how frequently each of the V and J gene occurs in the dataset; shows joint V and J gene
distributions as well; here the report can also be split by a label of interest</p></li>
<li><p><a class="reference internal" href="../yaml_specs/reports.html#labeldist"><span class="std std-ref">LabelDist</span></a>: shows the distribution for a list of labels of interest (e.g., disease, batch, age, sex); useful to
see if the dataset is balanced or if some unexpected values are present</p></li>
</ul>
<p>Some useful reports specifically for repertoire datasets (with one repertoire per patient):</p>
<ul class="simple">
<li><p><a class="reference internal" href="../yaml_specs/reports.html#repertoireclonotypesummary"><span class="std std-ref">RepertoireClonotypeSummary</span></a>: shows the number of distinct clonotypes per repertoire; the clonotype counts can be
colored by label of interest to see if there are any differences in the number of clonotypes per label.</p></li>
</ul>
<p>Additional useful reports to be applied on encoded datasets:</p>
<ul class="simple">
<li><p><a class="reference internal" href="../yaml_specs/reports.html#featurecomparison"><span class="std std-ref">FeatureComparison</span></a>: splits the data by the label of interests and plots their average values per label as a
scatterplot, allowing us to visually inspect which features differ the most between the label values.</p></li>
<li><p><a class="reference internal" href="../yaml_specs/reports.html#dimensionalityreduction"><span class="std std-ref">DimensionalityReduction</span></a>: uses a dimensionality reduction method (e.g., PCA, tSNE, UMAP) to visualize the encoded
dataset and optionally colors the points (each corresponding to one example from the dataset) by label(s) of interest.</p></li>
</ul>
<p>Encodings useful for exploratory analyses:</p>
<ul class="simple">
<li><p><a class="reference internal" href="../yaml_specs/encodings.html#kmerfrequency"><span class="std std-ref">KmerFrequency</span></a> encoder: represents an example by the frequency of subsequences of the given length; see the documentation
for additional configuration options;</p></li>
</ul>
<p>The analysis specification to run these reports:</p>
<details class="summary-step1-exploratory-analysis-yaml">
<summary>step1_exploratory_analysis.yaml</summary><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">definitions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">datasets</span><span class="p">:</span>
<span class="w">    </span><span class="nt">dataset</span><span class="p">:</span>
<span class="w">      </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AIRR</span>
<span class="w">      </span><span class="nt">params</span><span class="p">:</span><span class="w"> </span><span class="c1"># make sure the paths to the downloaded dataset files are correct</span>
<span class="w">        </span><span class="nt">dataset_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">./dataset/dataset.yaml</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">./dataset/metadata.csv</span>

<span class="w">  </span><span class="nt">encodings</span><span class="p">:</span>
<span class="w">    </span><span class="nt">3mer_freq</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">KmerFrequency</span>

<span class="w">  </span><span class="nt">reports</span><span class="p">:</span>

<span class="w">    </span><span class="nt">aa_dist</span><span class="p">:</span>
<span class="w">      </span><span class="nt">AminoAcidFrequencyDistribution</span><span class="p">:</span>
<span class="w">        </span><span class="nt">relative_frequency</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">        </span><span class="nt">split_by_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">        </span><span class="nt">label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">disease</span>
<span class="w">        </span><span class="nt">alignment</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span>

<span class="w">    </span><span class="nt">label_overlap</span><span class="p">:</span>
<span class="w">      </span><span class="nt">LabelOverlap</span><span class="p">:</span>
<span class="w">        </span><span class="nt">column_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">disease</span>
<span class="w">        </span><span class="nt">row_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">batch</span>

<span class="w">    </span><span class="nt">seq_len_dist_for_disease</span><span class="p">:</span>
<span class="w">      </span><span class="nt">SequenceLengthDistribution</span><span class="p">:</span>
<span class="w">        </span><span class="nt">sequence_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AMINO_ACID</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span>
<span class="w">        </span><span class="nt">label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">disease</span>
<span class="w">        </span><span class="nt">split_by_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">        </span><span class="nt">plot_frequencies</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>

<span class="w">    </span><span class="nt">seq_len_dist_for_batch</span><span class="p">:</span>
<span class="w">      </span><span class="nt">SequenceLengthDistribution</span><span class="p">:</span>
<span class="w">        </span><span class="nt">sequence_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AMINO_ACID</span>
<span class="w">        </span><span class="nt">region_type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">IMGT_CDR3</span>
<span class="w">        </span><span class="nt">label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">batch</span>
<span class="w">        </span><span class="nt">split_by_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">        </span><span class="nt">plot_frequencies</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>

<span class="w">    </span><span class="nt">gene_dist_disease</span><span class="p">:</span>
<span class="w">      </span><span class="nt">VJGeneDistribution</span><span class="p">:</span>
<span class="w">        </span><span class="nt">split_by_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">        </span><span class="nt">label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">disease</span>

<span class="w">    </span><span class="nt">gene_dist_batch</span><span class="p">:</span>
<span class="w">      </span><span class="nt">VJGeneDistribution</span><span class="p">:</span>
<span class="w">        </span><span class="nt">split_by_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">        </span><span class="nt">label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">batch</span>

<span class="w">    </span><span class="nt">clonotype_summary</span><span class="p">:</span>
<span class="w">      </span><span class="nt">RepertoireClonotypeSummary</span><span class="p">:</span>
<span class="w">        </span><span class="nt">color_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">batch</span>
<span class="w">        </span><span class="nt">facet_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">disease</span>

<span class="w">    </span><span class="nt">feature_comparison</span><span class="p">:</span>
<span class="w">      </span><span class="nt">FeatureComparison</span><span class="p">:</span>
<span class="w">        </span><span class="nt">comparison_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">disease</span>
<span class="w">        </span><span class="nt">column_grouping_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">batch</span>
<span class="w">        </span><span class="nt">opacity</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.6</span>
<span class="w">        </span><span class="nt">show_error_bar</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>

<span class="w">    </span><span class="nt">dim_red</span><span class="p">:</span>
<span class="w">      </span><span class="nt">DimensionalityReduction</span><span class="p">:</span>
<span class="w">        </span><span class="nt">labels</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">disease</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">batch</span><span class="p p-Indicator">]</span>
<span class="w">        </span><span class="nt">dim_red_method</span><span class="p">:</span>
<span class="w">          </span><span class="nt">KernelPCA</span><span class="p">:</span>
<span class="w">            </span><span class="nt">n_components</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="w">            </span><span class="nt">kernel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">rbf</span>

<span class="nt">instructions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">exploratory_analysis</span><span class="p">:</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ExploratoryAnalysis</span>
<span class="w">    </span><span class="nt">analyses</span><span class="p">:</span>
<span class="w">      </span><span class="nt">raw_data_analysis</span><span class="p">:</span>
<span class="w">        </span><span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset</span>
<span class="w">        </span><span class="nt">reports</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">aa_dist</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">label_overlap</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">seq_len_dist_for_disease</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">seq_len_dist_for_batch</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">gene_dist_disease</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">gene_dist_batch</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">clonotype_summary</span><span class="p p-Indicator">]</span>
<span class="w">      </span><span class="nt">encoded_data_analysis</span><span class="p">:</span>
<span class="w">        </span><span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset</span>
<span class="w">        </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3mer_freq</span>
<span class="w">        </span><span class="nt">reports</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">feature_comparison</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">dim_red</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
</details><p>Run the exploratory analysis from the command line:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>immune-ml<span class="w"> </span>step1_exploratory_analysis.yaml<span class="w"> </span>./exploratory_analysis_results/
</pre></div>
</div>
<p>Some of the results on the example dataset are shown below:</p>
<p>Amino acid frequency distribution report per IMGT position, split by ‘disease’ label:</p>
<img alt="Amino acid frequency distribution report" src="../_images/aa_freq_dist.png" />
<p>V gene distribution across repertoires split by the ‘disease’ label:</p>
<img alt="V gene usage across repertoires" src="../_images/V_gene_dist.png" />
<p>Comparison of feature values (average 3-mer frequencies per repertoire) across disease and batch labels:</p>
<img alt="Comparison of feature values across disease and batch labels" src="../_images/feature_val_comp.png" />
<p>Visualization of the dataset after 3-mer frequency encoding and dimensionality reduction with repertoires (examples)
colored by the ‘disease’ label:</p>
<img alt="Visualization of a dataset after 3-mer frequency encoding and dimensionality reduction with examples colored by disease label" src="../_images/dim_red_with_label.png" />
</section>
<section id="step-2-make-a-smaller-dataset-to-try-out-the-ml-analysis-first">
<h2>Step 2: Make a smaller dataset to try out the ML analysis first<a class="headerlink" href="#step-2-make-a-smaller-dataset-to-try-out-the-ml-analysis-first" title="Link to this heading">¶</a></h2>
<p>It is often a good idea to make a smaller dataset to try out the analysis first. This can be done by using the
Subsampling instruction.</p>
<details class="summary-step2-make-small-dataset-yaml">
<summary>step2_make_small_dataset.yaml</summary><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">definitions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">datasets</span><span class="p">:</span>
<span class="w">    </span><span class="nt">dataset</span><span class="p">:</span>
<span class="w">      </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AIRR</span>
<span class="w">      </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">dataset_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">./dataset/dataset.yaml</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">./dataset/metadata.csv</span>

<span class="nt">instructions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">subsampling</span><span class="p">:</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Subsampling</span>
<span class="w">    </span><span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset</span>
<span class="w">    </span><span class="nt">subsampled_dataset_sizes</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">75</span><span class="p p-Indicator">]</span><span class="w"> </span><span class="c1"># keep only 75 randomly selected repertoires</span>
<span class="w">    </span><span class="nt">subsampled_repertoire_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">50</span><span class="w"> </span><span class="c1"># keep only 50 sequences per repertoire</span>
</pre></div>
</div>
</details><p>Run this specification from the command line:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>immune-ml<span class="w"> </span>step2_make_small_dataset.yaml<span class="w"> </span>./small_dataset/
</pre></div>
</div>
<p>To explore the results, see <cite>./small_dataset/index.html</cite>. The small dataset in AIRR format is then located
at <cite>./small_dataset/subsampling/dataset_75_subsampled_1/exported/airr/</cite>.</p>
</section>
<section id="step-3-run-the-ml-analysis-on-a-smaller-dataset-revise-the-analysis-as-needed">
<h2>Step 3: Run the ML analysis on a smaller dataset &amp; revise the analysis as needed<a class="headerlink" href="#step-3-run-the-ml-analysis-on-a-smaller-dataset-revise-the-analysis-as-needed" title="Link to this heading">¶</a></h2>
<p>Here we define one example of immuneML analysis to build a repertoire classifier to distinguish between diseased and
healthy. We try out different ML methods and encodings and choose the “best” one using nested cross validation. For
more details on how this procedure is done in immuneML, see
<a class="reference internal" href="how_to_train_and_assess_a_receptor_or_repertoire_classifier.html#how-to-train-and-assess-a-receptor-or-repertoire-level-ml-classifier"><span class="std std-ref">How to train and assess a receptor or repertoire-level ML classifier</span></a>.
For a detailed list of data encodings and ML methods, see <a class="reference internal" href="../specification.html#yaml-specification"><span class="std std-ref">YAML specification</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The results we obtain here are not so interesting per se, as we run the analysis on a subset of data. The
aim is to make sure the analysis runs without issues and that all the parameters of interest (e.g., all
visualizations and ML-encoding combinations) have been provided.</p>
</div>
<p>Here is an example of the specification:</p>
<details class="summary-step3-small-ml-analysis-yaml">
<summary>step3_small_ML_analysis.yaml</summary><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">definitions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">datasets</span><span class="p">:</span>
<span class="w">    </span><span class="nt">dataset</span><span class="p">:</span>
<span class="w">      </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AIRR</span>
<span class="w">      </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">dataset_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">./small_dataset/subsampling/dataset_75_subsampled_1/exported/airr/dataset.yaml</span>
<span class="w">        </span><span class="nt">metadata_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">./small_dataset/subsampling/dataset_75_subsampled_1/exported/airr/metadata.csv</span>

<span class="w">  </span><span class="nt">encodings</span><span class="p">:</span>
<span class="w">    </span><span class="nt">3mer_freq</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">KmerFrequency</span>
<span class="w">    </span><span class="nt">esmc</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ESMC</span>

<span class="w">  </span><span class="nt">ml_methods</span><span class="p">:</span>
<span class="w">    </span><span class="nt">log_reg_l1</span><span class="p">:</span>
<span class="w">      </span><span class="nt">LogisticRegression</span><span class="p">:</span>
<span class="w">        </span><span class="nt">penalty</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">l1</span>
<span class="w">    </span><span class="nt">log_reg_l2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">LogisticRegression</span><span class="p">:</span>
<span class="w">        </span><span class="nt">penalty</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">l2</span>
<span class="w">    </span><span class="nt">random_forest</span><span class="p">:</span>
<span class="w">      </span><span class="nt">RandomForestClassifier</span><span class="p">:</span>
<span class="w">        </span><span class="nt">n_estimators</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">50</span>

<span class="w">  </span><span class="nt">reports</span><span class="p">:</span>
<span class="w">    </span><span class="nt">coefficients</span><span class="p">:</span><span class="w"> </span><span class="c1"># examine the coefficients of individual models (eg betas of logistic regression)</span>
<span class="w">      </span><span class="nt">Coefficients</span><span class="p">:</span>
<span class="w">        </span><span class="nt">coeff_to_plot</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">n_largest</span><span class="p p-Indicator">]</span>
<span class="w">        </span><span class="nt">n_largest</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">20</span><span class="p p-Indicator">]</span>

<span class="w">    </span><span class="nt">ml_setting_comparison</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">MLSettingsPerformance</span><span class="w"> </span><span class="c1"># plot performance of different ML settings</span>

<span class="w">    </span><span class="nt">performance_per_label</span><span class="p">:</span><span class="w"> </span><span class="c1"># performance of the model where examples are grouped by alternative label</span>
<span class="w">      </span><span class="nt">PerformancePerLabel</span><span class="p">:</span><span class="w"> </span><span class="c1"># it can be used to see if the model is learning alternative label instead of label of interest</span>
<span class="w">        </span><span class="nt">alternative_label</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">batch</span>
<span class="w">        </span><span class="nt">metric</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">balanced_accuracy</span>

<span class="w">    </span><span class="nt">roc_curve_summary</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ROCCurveSummary</span>

<span class="nt">instructions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">fit_and_eval_classifiers</span><span class="p">:</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TrainMLModel</span>
<span class="w">    </span><span class="nt">settings</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3mer_freq</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log_reg_l1</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3mer_freq</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log_reg_l2</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3mer_freq</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">random_forest</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">esmc</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log_reg_l1</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">esmc</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log_reg_l2</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">esmc</span>
<span class="w">        </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">random_forest</span>
<span class="w">    </span><span class="nt">assessment</span><span class="p">:</span><span class="w"> </span><span class="c1"># outer loop of the nested cross validation (for model assessment)</span>
<span class="w">      </span><span class="nt">split_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">random</span><span class="w"> </span><span class="c1"># randomly split to train+validation and test</span>
<span class="w">      </span><span class="nt">split_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">      </span><span class="nt">training_percentage</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span>
<span class="w">      </span><span class="nt">reports</span><span class="p">:</span>
<span class="w">        </span><span class="nt">models</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">coefficients</span><span class="p p-Indicator">]</span><span class="w"> </span><span class="c1"># plot coefficients of models trained on the train+validation data</span>
<span class="w">    </span><span class="nt">selection</span><span class="p">:</span><span class="w"> </span><span class="c1"># inner loop of the nested cross validation (for model selection)</span>
<span class="w">      </span><span class="nt">split_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">k_fold</span><span class="w"> </span><span class="c1"># do 5-fold CV starting from train+validation data [70% of the original dataset as defined in assessment]</span>
<span class="w">      </span><span class="nt">split_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">    </span><span class="nt">labels</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">disease</span><span class="p p-Indicator">]</span>
<span class="w">    </span><span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset</span>
<span class="w">    </span><span class="nt">metrics</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">auc</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">precision</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">recall</span><span class="p p-Indicator">]</span>
<span class="w">    </span><span class="nt">optimization_metric</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">balanced_accuracy</span>
<span class="w">    </span><span class="nt">reports</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">ml_setting_comparison</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">performance_per_label</span><span class="p p-Indicator">,</span><span class="w"> </span><span class="nv">roc_curve_summary</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
</details><p>Run this specification from the command line:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>immune-ml<span class="w"> </span>step3_small_ML_analysis.yaml<span class="w"> </span>./small_ML_result/
</pre></div>
</div>
<p>To explore the results, see <cite>./small_ML_result/index.html</cite>. In case something went wrong, all details can be found at
<cite>./small_ML_result/log.txt</cite>.</p>
</section>
<section id="step-4-run-the-analysis-on-the-full-dataset">
<h2>Step 4: Run the analysis on the full dataset<a class="headerlink" href="#step-4-run-the-analysis-on-the-full-dataset" title="Link to this heading">¶</a></h2>
<p>If everything worked out well so far (or if all necessary adjustments have been made), run the analysis on the full
dataset. The specification file is the same as in step 3, except that dataset file and metadata file should point to
the full dataset instead.</p>
<p>Here are some of the results on the full dataset:</p>
<ul class="simple">
<li><p>ROC curves with AUC annotated for all combinations of encoding + ML method listed on the test set:</p></li>
</ul>
<img alt="ROC curves with AUC for encodings and ML methods" src="../_images/roc_full.png" />
<p>As the disease-associated signal is a 3-mer (by the simulation setup), the best performing method is a
logistic regression with L1 penalty and with data encoded by 3-mer frequencies.</p>
<ul class="simple">
<li><p>Since there is also batch information in the dataset, it is possible to check if the performance of the
models differs for different batches (the results shown here are for the test set):</p></li>
</ul>
<img alt="Performance on label disease split by the batch for test set" src="../_images/bacc_per_batch.png" />
</section>
<section id="additional-information">
<h2>Additional information<a class="headerlink" href="#additional-information" title="Link to this heading">¶</a></h2>
<p>The simulated dataset used in this tutorial was created using LIgO tool from immuneML. It uses OLGA to generate
naive immune receptor sequences and simulates immune events by either implanting or filtering sequences based on
user-specified conditions. For more information on LIgO and simulations, see <a class="reference internal" href="ligo_simulation_tutorials.html#dataset-simulation-with-ligo"><span class="std std-ref">Dataset simulation with LIgO</span></a>.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="how_to_import_the_data_to_immuneML.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">How to import data into immuneML</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../tutorials.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Tutorials</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2025, Milena Pavlovic, Lonneke Scheffer, Keshav Motwani, Victor Greiff, Geir Kjetil Sandve
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Analyzing Your Own Dataset</a><ul>
<li><a class="reference internal" href="#step-1-explore-the-dataset">Step 1: Explore the dataset</a></li>
<li><a class="reference internal" href="#step-2-make-a-smaller-dataset-to-try-out-the-ml-analysis-first">Step 2: Make a smaller dataset to try out the ML analysis first</a></li>
<li><a class="reference internal" href="#step-3-run-the-ml-analysis-on-a-smaller-dataset-revise-the-analysis-as-needed">Step 3: Run the ML analysis on a smaller dataset &amp; revise the analysis as needed</a></li>
<li><a class="reference internal" href="#step-4-run-the-analysis-on-the-full-dataset">Step 4: Run the analysis on the full dataset</a></li>
<li><a class="reference internal" href="#additional-information">Additional information</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=f1196aca"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>