<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content="summary" name="twitter:card" />
<meta content="&#64;immuneml" name="twitter:site" />
<meta content="immuneML dev docs: add a new encoding" name="twitter:title" />
<meta content="See how to add a new encoding to the immuneML platform." name="twitter:description" />
<meta content="https://docs.immuneml.uio.no/_images/extending_immuneML.png" name="twitter:image" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="How to add a new machine learning method" href="how_to_add_new_ML_method.html" /><link rel="prev" title="Set up immuneML for development" href="install_for_development.html" />
        <link rel="canonical" href="https://docs.immuneml.uio.no/developer_docs/how_to_add_new_encoding.html" />

    <link rel="shortcut icon" href="../_static/favicon.ico"/><!-- Generated with Sphinx 8.1.3 and Furo 2024.08.06 -->
        <title>How to add a new encoding - immuneML 3.0.18 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/immuneml.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">immuneML 3.0.18 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">immuneML 3.0.18 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 has-children"><a class="reference internal" href="../quickstart.html">Quickstart</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Quickstart</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/galaxy_simple.html">Quickstart: Galaxy with button-based tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/galaxy_yaml.html">Quickstart: Galaxy with YAML-based tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/cli_yaml.html">Quickstart: command-line interface with YAML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/simulation_quickstart.html">LIgO simulation quickstart</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../installation.html">Installing immuneML</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Installing immuneML</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation/install_with_package_manager.html">Install immuneML with a package manager</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/installation_docker.html">Setting up immuneML with Docker</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/cloud.html">Running immuneML in the cloud</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../specification.html">YAML specification</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of YAML specification</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/how_to_specify_an_analysis_with_yaml.html">How to specify an analysis with YAML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/datasets.html">Dataset parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/encodings.html">Encoding parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/ml_methods.html">ML method parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/reports.html">Report parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/preprocessings.html">Preprocessing parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/simulation.html">Simulation parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/instructions.html">Instruction parameters</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/analyze_dataset.html">Analyzing Your Own Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_import_the_data_to_immuneML.html">How to import data into immuneML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_generate_a_random_repertoire_dataset.html">How to generate a dataset with random sequences</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../tutorials/ligo_simulation_tutorials.html">Dataset simulation with LIgO</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of Dataset simulation with LIgO</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/ligo_simulation_yaml.html">YAML specification of the LigoSim instruction for introducing immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/how_to_simulate_co-occuring_signals.html">How to simulate co-occuring immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/how_to_simulate_paired_chain_data.html">Paired chain simulations in LIgO</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/simulation_with_custom_signal_functions.html">Simulation with custom signal functions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_train_and_assess_a_receptor_or_repertoire_classifier.html">How to train and assess a receptor or repertoire-level ML classifier</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_apply_to_new_data.html">How to apply previously trained ML models to a new dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_perform_exploratory_analysis.html">How to perform an exploratory data analysis</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../tutorials/motif_recovery.html">How to find motifs associated with disease or antigen binding state</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of How to find motifs associated with disease or antigen binding state</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/discover_motifs_precision_recall.html">Discovering positional motifs using precision and recall thresholds</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/discover_motifs_classifiers.html">Discovering motifs learned by classifiers</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/recovering_simulated_motifs.html">Recovering simulated immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/comparing_baseline_motifs.html">Comparing baseline motif frequencies in repertoires</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/clustering_tutorial.html">How to perform clustering analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_train_and_apply_gen_model.html">How to train a generative model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_combine_multiple_encodings.html">How to combine multiple encodings to represent a dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/getting_publication_ready_figures.html">How to get publication-ready figures</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../galaxy.html">immuneML &amp; Galaxy</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of immuneML &amp; Galaxy</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_intro.html">Introduction to Galaxy</a></li>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_tools.html">immuneML Galaxy tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_immunologist_friendly.html">ML basics: Training classifiers with the simplified Galaxy interface</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../usecases.html">Use case examples</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of Use case examples</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../usecases/emerson_reproduction.html">Manuscript use case 1: Reproduction of a published study inside immuneML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/extendability_use_case.html">Manuscript use case 2: Extending immuneML with a deep learning component for predicting antigen specificity of paired receptor data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/benchmarking_use_case.html">Manuscript use case 3: Benchmarking ML methods on ground-truth synthetic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immcantation_use_case.html">Integration use case: post-analysis of sequences with Immcantation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immunarch_use_case.html">Integration use case: post-analysis of sequences with immunarch</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immunesim_use_case.html">Integration use case: Performing analysis on immuneSIM-generated repertoires</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../troubleshooting.html">Troubleshooting</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of Troubleshooting</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../developer_docs.html">Developer documentation</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of Developer documentation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="info_new_developers.html">Information for new developers</a></li>
<li class="toctree-l2"><a class="reference internal" href="install_for_development.html">Set up immuneML for development</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">How to add a new encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_add_new_ML_method.html">How to add a new machine learning method</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_add_new_report.html">How to add a new report</a></li>
<li class="toctree-l2"><a class="reference internal" href="how_to_add_new_preprocessing.html">How to add a new preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_model.html">immuneML data model</a></li>
<li class="toctree-l2"><a class="reference internal" href="execution_flow.html">immuneML execution flow</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/developer_docs/how_to_add_new_encoding.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="how-to-add-a-new-encoding">
<h1>How to add a new encoding<a class="headerlink" href="#how-to-add-a-new-encoding" title="Link to this heading">¶</a></h1>
<section id="adding-an-example-encoder-to-the-immuneml-codebase">
<h2>Adding an example encoder to the immuneML codebase<a class="headerlink" href="#adding-an-example-encoder-to-the-immuneml-codebase" title="Link to this heading">¶</a></h2>
<p>This tutorial describes how to add a new encoder to immuneML. That task is performed by implementing a subclass of
<code class="xref py py-obj docutils literal notranslate"><span class="pre">DatasetEncoder</span></code> shown here on the example of a simple encoder. We highly
recommend completing this tutorial to get a better understanding of the immuneML
interfaces before continuing to <a class="reference internal" href="#implementing-a-new-encoder"><span class="std std-ref">implement your own encoder</span></a>.</p>
<section id="step-by-step-tutorial">
<h3>Step-by-step tutorial<a class="headerlink" href="#step-by-step-tutorial" title="Link to this heading">¶</a></h3>
<p>For this tutorial, we provide a <code class="code docutils literal notranslate"><span class="pre">SillyEncoder</span></code> (<a class="reference download internal" download="" href="../_downloads/23b5c43e0cf50264eaeff19d815ecfa7/SillyEncoder.py"><code class="xref download docutils literal notranslate"><span class="pre">download</span> <span class="pre">here</span></code></a> or view below), in order to test adding a new Encoder file to immuneML.
This encoder ignores the data of the input examples, and generates a few random features per example.</p>
<details class="summary-sillyencoder-py">
<summary>SillyEncoder.py</summary><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">from</span> <span class="nn">immuneML.data_model.EncodedData</span> <span class="kn">import</span> <span class="n">EncodedData</span>
<span class="kn">from</span> <span class="nn">immuneML.data_model.SequenceParams</span> <span class="kn">import</span> <span class="n">RegionType</span><span class="p">,</span> <span class="n">Chain</span>
<span class="kn">from</span> <span class="nn">immuneML.data_model.SequenceSet</span> <span class="kn">import</span> <span class="n">ReceptorSequence</span>
<span class="kn">from</span> <span class="nn">immuneML.data_model.datasets.Dataset</span> <span class="kn">import</span> <span class="n">Dataset</span>
<span class="kn">from</span> <span class="nn">immuneML.data_model.datasets.ElementDataset</span> <span class="kn">import</span> <span class="n">SequenceDataset</span><span class="p">,</span> <span class="n">ReceptorDataset</span>
<span class="kn">from</span> <span class="nn">immuneML.data_model.datasets.RepertoireDataset</span> <span class="kn">import</span> <span class="n">RepertoireDataset</span>
<span class="kn">from</span> <span class="nn">immuneML.encodings.DatasetEncoder</span> <span class="kn">import</span> <span class="n">DatasetEncoder</span>
<span class="kn">from</span> <span class="nn">immuneML.encodings.EncoderParams</span> <span class="kn">import</span> <span class="n">EncoderParams</span>
<span class="kn">from</span> <span class="nn">immuneML.util.EncoderHelper</span> <span class="kn">import</span> <span class="n">EncoderHelper</span>
<span class="kn">from</span> <span class="nn">immuneML.util.ParameterValidator</span> <span class="kn">import</span> <span class="n">ParameterValidator</span>


<span class="k">class</span> <span class="nc">SillyEncoder</span><span class="p">(</span><span class="n">DatasetEncoder</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This SillyEncoder class is a placeholder for a real encoder.</span>
<span class="sd">    It computes a set of random numbers as features for a given dataset.</span>

<span class="sd">    **Dataset type:**</span>

<span class="sd">    - SequenceDatasets</span>

<span class="sd">    - ReceptorDatasets</span>

<span class="sd">    - RepertoireDatasets</span>


<span class="sd">    **Specification arguments:**</span>

<span class="sd">    - random_seed (int): The random seed for generating random features.</span>

<span class="sd">    - embedding_len (int): The number of random features to generate per example.</span>


<span class="sd">    **YAML specification:**</span>

<span class="sd">    .. indent with spaces</span>
<span class="sd">    .. code-block:: yaml</span>

<span class="sd">        definitions:</span>
<span class="sd">            encodings:</span>
<span class="sd">                my_silly_encoder:</span>
<span class="sd">                    Silly: # name of the class (without &#39;Encoder&#39; suffix)</span>
<span class="sd">                        random_seed: 1</span>
<span class="sd">                        embedding_len: 5</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">random_seed</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">embedding_len</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="c1"># The encoder name contains the user-defined name for the encoder. It may be used by some reports.</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>

        <span class="c1"># All user parameters are set here.</span>
        <span class="c1"># Default parameters must not be defined in the Encoder class, but in a default parameters file.</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">random_seed</span> <span class="o">=</span> <span class="n">random_seed</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">embedding_len</span> <span class="o">=</span> <span class="n">embedding_len</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">build_object</span><span class="p">(</span><span class="n">dataset</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">params</span><span class="p">):</span>
        <span class="c1"># build_object is called early in the immuneML run, before the analysis takes place.</span>
        <span class="c1"># Its purpose is to fail early when a class is called incorrectly (checking parameters and dataset),</span>
        <span class="c1"># and provide user-friendly error messages.</span>

        <span class="c1"># ParameterValidator contains many utility functions for checking user parameters</span>
        <span class="n">ParameterValidator</span><span class="o">.</span><span class="n">assert_type_and_value</span><span class="p">(</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;random_seed&#39;</span><span class="p">],</span> <span class="nb">int</span><span class="p">,</span> <span class="n">SillyEncoder</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span> <span class="s1">&#39;random_seed&#39;</span><span class="p">,</span> <span class="n">min_inclusive</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ParameterValidator</span><span class="o">.</span><span class="n">assert_type_and_value</span><span class="p">(</span><span class="n">params</span><span class="p">[</span><span class="s1">&#39;embedding_len&#39;</span><span class="p">],</span> <span class="nb">int</span><span class="p">,</span> <span class="n">SillyEncoder</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span> <span class="s1">&#39;embedding_len&#39;</span><span class="p">,</span> <span class="n">min_inclusive</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_inclusive</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

        <span class="c1"># An error should be thrown if the dataset type is incompatible with the Encoder.</span>
        <span class="c1"># If different sub-classes are defined for each dataset type (e.g., OneHotRepertoireEncoder),</span>
        <span class="c1"># an instance of the dataset-specific class must be returned here.</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">SequenceDataset</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">ReceptorDataset</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">RepertoireDataset</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">SillyEncoder</span><span class="p">(</span><span class="o">**</span><span class="n">params</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;SillyEncoder is only defined for dataset types SequenceDataset, ReceptorDataset or RepertoireDataset&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">encode</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">params</span><span class="p">:</span> <span class="n">EncoderParams</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dataset</span><span class="p">:</span>
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">random_seed</span><span class="p">)</span>

        <span class="c1"># Generate the design matrix from the sequence dataset</span>
        <span class="n">encoded_examples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_encoded_examples</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>

        <span class="c1"># EncoderHelper contains some utility functions, including this function for encoding the labels</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="n">EncoderHelper</span><span class="o">.</span><span class="n">encode_dataset_labels</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">params</span><span class="o">.</span><span class="n">label_config</span><span class="p">,</span> <span class="n">params</span><span class="o">.</span><span class="n">encode_labels</span><span class="p">)</span>

        <span class="c1"># Each feature is represented by some meaningful name</span>
        <span class="n">feature_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;random_number_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">embedding_len</span><span class="p">)]</span>

        <span class="n">encoded_data</span> <span class="o">=</span> <span class="n">EncodedData</span><span class="p">(</span><span class="n">examples</span><span class="o">=</span><span class="n">encoded_examples</span><span class="p">,</span>
                                   <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span>
                                   <span class="n">example_ids</span><span class="o">=</span><span class="n">dataset</span><span class="o">.</span><span class="n">get_example_ids</span><span class="p">(),</span>
                                   <span class="n">feature_names</span><span class="o">=</span><span class="n">feature_names</span><span class="p">,</span>
                                   <span class="n">encoding</span><span class="o">=</span><span class="n">SillyEncoder</span><span class="o">.</span><span class="vm">__name__</span><span class="p">)</span> <span class="c1"># When using dataset-specific encoders,</span>
                                                                   <span class="c1"># make sure to use the general encoder name here</span>
                                                                   <span class="c1"># (e.g., OneHotEncoder.__name__, not OneHotSequenceEncoder.__name__)</span>

        <span class="n">encoded_dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
        <span class="n">encoded_dataset</span><span class="o">.</span><span class="n">encoded_data</span> <span class="o">=</span> <span class="n">encoded_data</span>

        <span class="k">return</span> <span class="n">encoded_dataset</span>

    <span class="k">def</span> <span class="nf">_get_encoded_examples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataset</span><span class="p">:</span> <span class="n">Dataset</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">SequenceDataset</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_encoded_sequences</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">ReceptorDataset</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_encoded_receptors</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">RepertoireDataset</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_encoded_repertoires</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_encoded_sequences</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataset</span><span class="p">:</span> <span class="n">SequenceDataset</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">:</span>

        <span class="c1"># for sequence dataset, the data can be access as a list of ReceptorSequence objects:</span>
        <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">dataset</span><span class="o">.</span><span class="n">get_data</span><span class="p">():</span>
            <span class="c1"># Each sequence is a ReceptorSequence object.</span>
            <span class="c1"># Different properties of the sequence can be retrieved here, examples:</span>
            <span class="n">identifier</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">sequence_id</span>
            <span class="n">aa_seq</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">sequence_aa</span>
            <span class="n">v_gene</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">v_call</span>
            <span class="n">j_gene</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">j_call</span>

        <span class="c1"># alternatively, data can be accessed through the AIRRSequenceSet object</span>
        <span class="c1"># (a bionumpy dataclass, similar to pandas dataframes):</span>

        <span class="n">all_cdr3_aa</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cdr3_aa</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># a list of strings of all sequences in the dataset</span>
        <span class="n">all_v_call</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">v_call</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># a list of strings of all v genes in the dataset</span>

        <span class="c1"># it can also be accessed as a bionumpy dataclass object:</span>
        <span class="n">cdr3_aa</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cdr3_aa</span>  <span class="c1"># useful for faster computations</span>

        <span class="c1"># in this example, we just return a random matrix of n_examples x n_features</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">element_count</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">embedding_len</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_encoded_receptors</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataset</span><span class="p">:</span> <span class="n">ReceptorDataset</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">:</span>

        <span class="c1"># for receptor dataset, the data can be access as a list of Receptor objects:</span>
        <span class="k">for</span> <span class="n">receptor</span> <span class="ow">in</span> <span class="n">dataset</span><span class="o">.</span><span class="n">get_data</span><span class="p">():</span>
            <span class="c1"># A Receptor contains two paired ReceptorSequence objects</span>
            <span class="n">identifier</span> <span class="o">=</span> <span class="n">receptor</span><span class="o">.</span><span class="n">receptor_id</span>
            <span class="n">chain1</span><span class="p">,</span> <span class="n">chain2</span> <span class="o">=</span> <span class="n">receptor</span><span class="o">.</span><span class="n">get_chains</span><span class="p">()</span>
            <span class="n">sequence1</span> <span class="o">=</span> <span class="n">receptor</span><span class="o">.</span><span class="n">get_chain</span><span class="p">(</span><span class="n">chain1</span><span class="p">)</span>
            <span class="n">sequence2</span> <span class="o">=</span> <span class="n">receptor</span><span class="o">.</span><span class="n">get_chain</span><span class="p">(</span><span class="n">chain2</span><span class="p">)</span>

            <span class="c1"># Properties of the specific ReceptorSequences can be retrieved, examples:</span>
            <span class="n">aa_seq1</span> <span class="o">=</span> <span class="n">sequence1</span><span class="o">.</span><span class="n">sequence_aa</span> <span class="c1"># gets the amino acid sequence by default (alternative: nucleotide)</span>
            <span class="n">v_gene_seq1</span> <span class="o">=</span> <span class="n">sequence1</span><span class="o">.</span><span class="n">v_call</span> <span class="c1"># gets the v and j genes</span>
            <span class="n">j_gene_seq1</span> <span class="o">=</span> <span class="n">sequence1</span><span class="o">.</span><span class="n">j_call</span>

        <span class="c1"># data can be also accessed through the AIRRSequenceSet object</span>
        <span class="c1"># (a bionumpy dataclass, similar to pandas dataframes)</span>
        <span class="c1"># this is useful when speed is important</span>

        <span class="c1"># all cdr3 aa sequences of the beta chain</span>
        <span class="n">cdr3_aa_beta</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cdr3_aa</span><span class="p">[</span><span class="n">dataset</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">locus</span> <span class="o">==</span> <span class="n">Chain</span><span class="o">.</span><span class="n">BETA</span><span class="o">.</span><span class="n">to_string</span><span class="p">()]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="c1"># all cdr3 aa sequences of the alpha chain</span>
        <span class="n">cdr3_aa_alpha</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cdr3_aa</span><span class="p">[</span><span class="n">dataset</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">locus</span> <span class="o">==</span> <span class="n">Chain</span><span class="o">.</span><span class="n">ALPHA</span><span class="o">.</span><span class="n">to_string</span><span class="p">()]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="c1"># here we just return a random matrix of n_examples x n_features</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">element_count</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">embedding_len</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_encoded_repertoires</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataset</span><span class="p">:</span> <span class="n">RepertoireDataset</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">:</span>
        <span class="n">encoded_repertoires</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">repertoire</span> <span class="ow">in</span> <span class="n">dataset</span><span class="o">.</span><span class="n">get_data</span><span class="p">():</span>
            <span class="c1"># Each repertoire is a Repertoire object.</span>
            <span class="c1"># Receptor sequences from the repertoire objects can be retrieved as objects for the given region type:</span>
            <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">repertoire</span><span class="o">.</span><span class="n">sequences</span><span class="p">(</span><span class="n">region_type</span><span class="o">=</span><span class="n">RegionType</span><span class="o">.</span><span class="n">IMGT_CDR3</span><span class="p">):</span>
                <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">ReceptorSequence</span><span class="p">)</span> <span class="c1"># each sequence is an object of ReceptorSequence class</span>
                <span class="n">seq_id</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">sequence_id</span>
                <span class="n">aa_seq</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">sequence_aa</span>
                <span class="n">v_call</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">v_call</span>

            <span class="c1"># alternatively, the sequences can be retrieved as AIRRSequenceSet objects (bionumpy dataclasses,</span>
            <span class="c1"># similar to pandas dataframes):</span>
            <span class="n">all_aa_sequences</span> <span class="o">=</span> <span class="n">repertoire</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cdr3_aa</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># a list of strings of all sequences in the repertoire</span>
            <span class="n">all_aa_sequences</span> <span class="o">=</span> <span class="n">repertoire</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cdr3_aa</span>  <span class="c1"># a bionumpy dataclass object containing all sequences in the repertoire</span>

            <span class="c1"># other properties can also be accessed in the same way:</span>
            <span class="n">v_calls</span> <span class="o">=</span> <span class="n">repertoire</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">v_call</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># a list of strings of all v genes in the repertoire</span>

            <span class="c1"># In this encoding, repertoire information is ignored, random features are generated</span>
            <span class="n">random_encoding</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">embedding_len</span><span class="p">)</span>
            <span class="n">encoded_repertoires</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random_encoding</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">encoded_repertoires</span><span class="p">)</span>
</pre></div>
</div>
</details><ol class="arabic simple">
<li><p>Add a new <a class="reference external" href="https://packaging.python.org/en/latest/tutorials/packaging-projects/">Python package</a> to the <code class="xref py py-obj docutils literal notranslate"><span class="pre">encodings</span></code> package.
This means: a new folder (with meaningful name) containing an empty <code class="code docutils literal notranslate"><span class="pre">__init__.py</span></code> file.</p></li>
<li><p>Add a new encoder class to the package. The new class should inherit from the base class <code class="xref py py-obj docutils literal notranslate"><span class="pre">DatasetEncoder</span></code>.
The name of the class should end with ‘Encoder’, and when calling this class in the YAML specification, the ‘Encoder’ suffix is omitted.
In the test example, the class is called <code class="code docutils literal notranslate"><span class="pre">SillyEncoder</span></code>, which would be referred to as <code class="code docutils literal notranslate"><span class="pre">Silly</span></code> in the YAML specification.</p></li>
<li><p>If the encoder has any default parameters, they should be added in a default parameters YAML file. This file should be added to the folder <code class="code docutils literal notranslate"><span class="pre">config/default_params/encodings</span></code>.
The default parameters file is automatically discovered based on the name of the class using the base name (without ‘Encoder’ suffix) converted to snake case, and with an added ‘_params.yaml’ suffix.
For the <code class="code docutils literal notranslate"><span class="pre">SillyEncoder</span></code>, this is <code class="code docutils literal notranslate"><span class="pre">silly_params.yaml</span></code>, which could for example contain the following:</p></li>
</ol>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="w">     </span><span class="nt">random_seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">     </span><span class="nt">embedding_len</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>

<span class="l l-Scalar l-Scalar-Plain">In rare cases where classes have unconventional names that do not translate well to CamelCase (e.g., MiXCR, VDJdb),</span>
<span class="l l-Scalar l-Scalar-Plain">this needs to be accounted for in :py:meth:`~immuneML.dsl.DefaultParamsLoader.convert_to_snake_case`.</span>
</pre></div>
</div>
<ol class="arabic simple">
<li><p><strong>Use the automated script</strong> <a class="reference external" href="https://github.com/uio-bmi/immuneML/blob/master/scripts/check_new_encoder.py">check_new_encoder.py</a> <strong>to test the newly added encoder.</strong>
This script will throw errors or warnings if the DatasetEncoder class implementation is incorrect or if files are put in the wrong place.
Example command to test the <code class="code docutils literal notranslate"><span class="pre">SillyEncoder</span></code> for sequence datasets:</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python3<span class="w"> </span>./scripts/check_new_encoder.py<span class="w"> </span>-e<span class="w"> </span>./immuneML/encodings/silly/SillyEncoder.py<span class="w"> </span>-d<span class="w"> </span>sequence
</pre></div>
</div>
<ol class="arabic simple">
<li><p>If a compatible ML method is already available, add the new encoder class to the list of compatible encoders returned by the
<code class="code docutils literal notranslate"><span class="pre">get_compatible_encoders()</span></code> method of the <code class="xref py py-obj docutils literal notranslate"><span class="pre">MLMethod</span></code> of interest.
See also <a class="reference internal" href="how_to_add_new_ML_method.html#adding-encoder-compatibility-to-an-ml-method"><span class="std std-ref">Adding encoder compatibility to an ML method</span></a>.</p></li>
</ol>
</section>
<section id="test-running-the-new-encoding-with-a-yaml-specification">
<h3>Test running the new encoding with a YAML specification<a class="headerlink" href="#test-running-the-new-encoding-with-a-yaml-specification" title="Link to this heading">¶</a></h3>
<p>If you want to use immuneML directly to test run your encoder, the YAML example below may be used.
This example analysis creates a randomly generated dataset, encodes the data using the <code class="code docutils literal notranslate"><span class="pre">SillyEncoder</span></code>
and exports the encoded data as a csv file.</p>
<details class="summary-test-run-silly-encoder-yaml">
<summary>test_run_silly_encoder.yaml</summary><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">definitions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">datasets</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_dataset</span><span class="p">:</span>
<span class="w">      </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">RandomSequenceDataset</span>
<span class="w">      </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">sequence_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span>
<span class="w">        </span><span class="nt">labels</span><span class="p">:</span>
<span class="w">          </span><span class="nt">binds_epitope</span><span class="p">:</span>
<span class="w">            </span><span class="nt">True</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.6</span>
<span class="w">            </span><span class="nt">False</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.4</span>

<span class="w">  </span><span class="nt">encodings</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_silly_encoder</span><span class="p">:</span>
<span class="w">      </span><span class="nt">Silly</span><span class="p">:</span>
<span class="w">        </span><span class="nt">random_seed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span>

<span class="w">  </span><span class="nt">reports</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_design_matrix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">DesignMatrixExporter</span>

<span class="nt">instructions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">my_instruction</span><span class="p">:</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ExploratoryAnalysis</span>
<span class="w">    </span><span class="nt">analyses</span><span class="p">:</span>
<span class="w">      </span><span class="nt">my_analysis_1</span><span class="p">:</span>
<span class="w">        </span><span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_dataset</span>
<span class="w">        </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_silly_encoder</span>
<span class="w">        </span><span class="nt">report</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_design_matrix</span>
<span class="w">        </span><span class="nt">labels</span><span class="p">:</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">binds_epitope</span>
</pre></div>
</div>
</details></section>
<section id="adding-a-unit-test-for-a-datasetencoder">
<h3>Adding a Unit test for a DatasetEncoder<a class="headerlink" href="#adding-a-unit-test-for-a-datasetencoder" title="Link to this heading">¶</a></h3>
<p>Add a unit test for the new SillyEncoder (<a class="reference download internal" download="" href="../_downloads/0b787a6450bde42b6a17dd2bb54ea2f0/_test_sillyEncoder.py"><code class="xref download docutils literal notranslate"><span class="pre">download</span></code></a> the example testfile or view below):</p>
<details class="summary-test-sillyencoder-py">
<summary>test_sillyEncoder.py</summary><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">unittest</span>

<span class="kn">from</span> <span class="nn">immuneML.caching.CacheType</span> <span class="kn">import</span> <span class="n">CacheType</span>
<span class="kn">from</span> <span class="nn">immuneML.data_model.EncodedData</span> <span class="kn">import</span> <span class="n">EncodedData</span>
<span class="kn">from</span> <span class="nn">immuneML.encodings.EncoderParams</span> <span class="kn">import</span> <span class="n">EncoderParams</span>
<span class="kn">from</span> <span class="nn">immuneML.environment.Constants</span> <span class="kn">import</span> <span class="n">Constants</span>
<span class="kn">from</span> <span class="nn">immuneML.environment.EnvironmentSettings</span> <span class="kn">import</span> <span class="n">EnvironmentSettings</span>
<span class="kn">from</span> <span class="nn">immuneML.environment.LabelConfiguration</span> <span class="kn">import</span> <span class="n">LabelConfiguration</span>
<span class="kn">from</span> <span class="nn">immuneML.encodings.silly.SillyEncoder</span> <span class="kn">import</span> <span class="n">SillyEncoder</span>
<span class="kn">from</span> <span class="nn">immuneML.environment.Label</span> <span class="kn">import</span> <span class="n">Label</span>
<span class="kn">from</span> <span class="nn">immuneML.simulation.dataset_generation.RandomDatasetGenerator</span> <span class="kn">import</span> <span class="n">RandomDatasetGenerator</span>


<span class="k">class</span> <span class="nc">TestSillyEncoder</span><span class="p">(</span><span class="n">unittest</span><span class="o">.</span><span class="n">TestCase</span><span class="p">):</span>

    <span class="k">def</span> <span class="nf">setUp</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="n">Constants</span><span class="o">.</span><span class="n">CACHE_TYPE</span><span class="p">]</span> <span class="o">=</span> <span class="n">CacheType</span><span class="o">.</span><span class="n">TEST</span><span class="o">.</span><span class="n">name</span>

    <span class="k">def</span> <span class="nf">_get_mock_sequence_dataset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">):</span>
        <span class="c1"># Create a mock SequenceDataset with 10 sequences of length 15,</span>
        <span class="c1"># and a label called &#39;binding&#39; with 50% chance of having status &#39;yes&#39; or &#39;no&#39;</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="n">RandomDatasetGenerator</span><span class="o">.</span><span class="n">generate_sequence_dataset</span><span class="p">(</span><span class="n">sequence_count</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                                                                   <span class="n">length_probabilities</span><span class="o">=</span><span class="p">{</span><span class="mi">15</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
                                                                   <span class="n">labels</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;binding&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;yes&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;no&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">}},</span>
                                                                   <span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">)</span>

        <span class="n">label_config</span> <span class="o">=</span> <span class="n">LabelConfiguration</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="n">Label</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;binding&quot;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;yes&quot;</span><span class="p">,</span> <span class="s2">&quot;no&quot;</span><span class="p">])])</span>

        <span class="k">return</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">label_config</span>

    <span class="k">def</span> <span class="nf">_get_mock_receptor_dataset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">):</span>
        <span class="c1"># Create a mock ReceptorDataset with 10 receptors with sequences of length 15,</span>
        <span class="c1"># and a label called &#39;binding&#39; with 50% chance of having status &#39;yes&#39; or &#39;no&#39;</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="n">RandomDatasetGenerator</span><span class="o">.</span><span class="n">generate_receptor_dataset</span><span class="p">(</span><span class="n">receptor_count</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                                                                   <span class="n">chain_1_length_probabilities</span><span class="o">=</span><span class="p">{</span><span class="mi">15</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
                                                                   <span class="n">chain_2_length_probabilities</span><span class="o">=</span><span class="p">{</span><span class="mi">15</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
                                                                   <span class="n">labels</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;binding&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;yes&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;no&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">}},</span>
                                                                   <span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">)</span>

        <span class="n">label_config</span> <span class="o">=</span> <span class="n">LabelConfiguration</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="n">Label</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;binding&quot;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;yes&quot;</span><span class="p">,</span> <span class="s2">&quot;no&quot;</span><span class="p">])])</span>

        <span class="k">return</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">label_config</span>

    <span class="k">def</span> <span class="nf">_get_mock_repertoire_dataset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">path</span><span class="p">):</span>
        <span class="c1"># Create a mock RepertoireDataset with 10 repertoires, each containing 50 sequences of length 15,</span>
        <span class="c1"># and a label called &#39;disease&#39; with 50% chance of having status &#39;yes&#39; or &#39;no&#39;</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="n">RandomDatasetGenerator</span><span class="o">.</span><span class="n">generate_repertoire_dataset</span><span class="p">(</span><span class="n">repertoire_count</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                                                           <span class="n">sequence_count_probabilities</span><span class="o">=</span><span class="p">{</span><span class="mi">50</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
                                                           <span class="n">sequence_length_probabilities</span><span class="o">=</span><span class="p">{</span><span class="mi">15</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
                                                           <span class="n">labels</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;disease&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;yes&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;no&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">}},</span>
                                                           <span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">)</span>

        <span class="n">label_config</span> <span class="o">=</span> <span class="n">LabelConfiguration</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="n">Label</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;disease&quot;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;yes&quot;</span><span class="p">,</span> <span class="s2">&quot;no&quot;</span><span class="p">])])</span>

        <span class="k">return</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">label_config</span>

    <span class="k">def</span> <span class="nf">test_silly_sequence_encoder</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">tmp_path</span> <span class="o">=</span> <span class="n">EnvironmentSettings</span><span class="o">.</span><span class="n">tmp_test_path</span> <span class="o">/</span> <span class="s2">&quot;silly_sequence/&quot;</span>
        <span class="n">sequence_dataset</span><span class="p">,</span> <span class="n">label_config</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_mock_sequence_dataset</span><span class="p">(</span><span class="n">tmp_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_test_silly_encoder</span><span class="p">(</span><span class="n">tmp_path</span><span class="p">,</span> <span class="n">sequence_dataset</span><span class="p">,</span> <span class="n">label_config</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">test_silly_receptor_encoder</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">tmp_path</span> <span class="o">=</span> <span class="n">EnvironmentSettings</span><span class="o">.</span><span class="n">tmp_test_path</span> <span class="o">/</span> <span class="s2">&quot;silly_receptor/&quot;</span>
        <span class="n">receptor_dataset</span><span class="p">,</span> <span class="n">label_config</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_mock_receptor_dataset</span><span class="p">(</span><span class="n">tmp_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_test_silly_encoder</span><span class="p">(</span><span class="n">tmp_path</span><span class="p">,</span> <span class="n">receptor_dataset</span><span class="p">,</span> <span class="n">label_config</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">test_silly_repertoire_encoder</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">tmp_path</span> <span class="o">=</span> <span class="n">EnvironmentSettings</span><span class="o">.</span><span class="n">tmp_test_path</span> <span class="o">/</span> <span class="s2">&quot;silly_repertoire/&quot;</span>
        <span class="n">receptor_dataset</span><span class="p">,</span> <span class="n">label_config</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_mock_repertoire_dataset</span><span class="p">(</span><span class="n">tmp_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_test_silly_encoder</span><span class="p">(</span><span class="n">tmp_path</span><span class="p">,</span> <span class="n">receptor_dataset</span><span class="p">,</span> <span class="n">label_config</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_test_silly_encoder</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tmp_path</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">label_config</span><span class="p">):</span>
        <span class="c1"># test getting a SillyEncoder from the build_object method</span>
        <span class="n">params</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;random_seed&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;embedding_len&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">}</span>
        <span class="n">encoder</span> <span class="o">=</span> <span class="n">SillyEncoder</span><span class="o">.</span><span class="n">build_object</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="o">**</span><span class="n">params</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assertIsInstance</span><span class="p">(</span><span class="n">encoder</span><span class="p">,</span> <span class="n">SillyEncoder</span><span class="p">)</span>

        <span class="c1"># test encoding data</span>
        <span class="n">encoded_dataset</span> <span class="o">=</span> <span class="n">encoder</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span>
                                         <span class="n">params</span><span class="o">=</span><span class="n">EncoderParams</span><span class="p">(</span><span class="n">result_path</span><span class="o">=</span><span class="n">tmp_path</span><span class="p">,</span>
                                                              <span class="n">label_config</span><span class="o">=</span><span class="n">label_config</span><span class="p">))</span>

        <span class="c1"># the result must be a Dataset (of the same subtype as the original dataset) with EncodedData attached</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assertIsInstance</span><span class="p">(</span><span class="n">encoded_dataset</span><span class="p">,</span> <span class="n">dataset</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assertIsInstance</span><span class="p">(</span><span class="n">encoded_dataset</span><span class="o">.</span><span class="n">encoded_data</span><span class="p">,</span> <span class="n">EncodedData</span><span class="p">)</span>

        <span class="c1"># testing the validity of the encoded data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assertEqual</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">get_example_ids</span><span class="p">(),</span> <span class="n">encoded_dataset</span><span class="o">.</span><span class="n">encoded_data</span><span class="o">.</span><span class="n">example_ids</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assertTrue</span><span class="p">((</span><span class="n">encoded_dataset</span><span class="o">.</span><span class="n">encoded_data</span><span class="o">.</span><span class="n">examples</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">())</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assertTrue</span><span class="p">((</span><span class="n">encoded_dataset</span><span class="o">.</span><span class="n">encoded_data</span><span class="o">.</span><span class="n">examples</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">())</span>

        <span class="c1"># don&#39;t forget to remove the temporary data</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">tmp_path</span><span class="p">)</span>
</pre></div>
</div>
</details><ol class="arabic simple">
<li><p>Add a new package to the <code class="code docutils literal notranslate"><span class="pre">test.encodings</span></code> package which matches the package name of your encoder code. In this case, the new package would be <code class="code docutils literal notranslate"><span class="pre">test.encodings.silly</span></code>.</p></li>
<li><p>To the new <code class="code docutils literal notranslate"><span class="pre">test.encodings.silly</span></code> package, add a new file named test_sillyEncoder.py.</p></li>
<li><p>Add a class <code class="code docutils literal notranslate"><span class="pre">TestSillyEncoder</span></code> that inherits <code class="code docutils literal notranslate"><span class="pre">unittest.TestCase</span></code> to the new file.</p></li>
<li><p>Add a function <code class="code docutils literal notranslate"><span class="pre">setUp()</span></code> to set up cache used for testing. This should ensure that the cache location will be set to <code class="code docutils literal notranslate"><span class="pre">EnvironmentSettings.tmp_test_path</span> <span class="pre">/</span> <span class="pre">&quot;cache/&quot;</span></code></p></li>
<li><p>Define one or more tests for the class and functions you implemented. For the SillyEncoder example, these have already been added. Note:</p>
<ul class="simple">
<li><p>It is recommended to at least test the output of the ‘encode’ method (ensure a valid EncodedData object with correct examples matrix is returned).</p></li>
<li><p>Make sure to add tests for <em>every</em> relevant dataset type. Tests for different dataset types may be split into several different classes/files if desired (e.g., test_oneHotReceptorEncoder.py, test_oneHotSequenceEncoder.py, …). For the SillyEncoder, all tests are in the same file.</p></li>
<li><p>Mock data is typically used to test new classes. Tip: the <code class="code docutils literal notranslate"><span class="pre">RandomDatasetGenerator</span></code> class can be used to generate Repertoire, Sequence or Receptor datasets with random sequences.</p></li>
<li><p>If you need to write data to a path (for example test datasets or results), use the following location: <code class="code docutils literal notranslate"><span class="pre">EnvironmentSettings.tmp_test_path</span> <span class="pre">/</span> <span class="pre">&quot;some_unique_foldername&quot;</span></code></p></li>
</ul>
</li>
</ol>
</section>
</section>
<section id="implementing-a-new-encoder">
<h2>Implementing a new encoder<a class="headerlink" href="#implementing-a-new-encoder" title="Link to this heading">¶</a></h2>
<p>This section describes tips and tricks for implementing your own new <code class="code docutils literal notranslate"><span class="pre">DatasetEncoder</span></code> from scratch.
Detailed instructions of how to implement each method, as well as some special cases, can be found in the
<code class="xref py py-obj docutils literal notranslate"><span class="pre">DatasetEncoder</span></code> base class.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>Coding conventions and tips</strong></p>
<ol class="arabic simple">
<li><p>Class names are written in CamelCase</p></li>
<li><p>Class methods are writte in snake_case</p></li>
<li><p>Abstract base classes <code class="code docutils literal notranslate"><span class="pre">MLMethod</span></code>, <code class="code docutils literal notranslate"><span class="pre">DatasetEncoder</span></code>, and <code class="code docutils literal notranslate"><span class="pre">Report</span></code>, define an interface for their inheriting subclasses. These classes contain abstract methods which should be implemented.</p></li>
<li><p>Class methods starting with _underscore are generally considered “private” methods, only to be called by the class itself. If a method is expected to be called from another class, the method name should not start with an underscore.</p></li>
<li><p>When familiarising yourself with existing code, we recommend focusing on public methods. Private methods are typically very unique to a class (internal class-specific calculations), whereas the public methods contain more general functionalities (e.g., returning a main result).</p></li>
<li><p>If your class should have any default parameters, they should be defined in a default parameters file under <code class="code docutils literal notranslate"><span class="pre">config/default_params/</span></code>.</p></li>
<li><p>Some utility classes are available in the <code class="code docutils literal notranslate"><span class="pre">util</span></code> package to provide useful functionalities. For example, <code class="xref py py-obj docutils literal notranslate"><span class="pre">ParameterValidator</span></code> can be used to check user input and generate error messages, or <code class="xref py py-obj docutils literal notranslate"><span class="pre">PathBuilder</span></code> can be used to add and remove folders.</p></li>
</ol>
</div>
<section id="encoders-for-different-dataset-types">
<h3>Encoders for different dataset types<a class="headerlink" href="#encoders-for-different-dataset-types" title="Link to this heading">¶</a></h3>
<p>Inside immuneML, three different types of datasets are considered: <code class="xref py py-obj docutils literal notranslate"><span class="pre">RepertoireDataset</span></code> for immune
repertoires, <code class="xref py py-obj docutils literal notranslate"><span class="pre">SequenceDataset</span></code> for single-chain immune
receptor sequences and <code class="xref py py-obj docutils literal notranslate"><span class="pre">ReceptorDataset</span></code> for paired sequences.
Encoding should be implemented separately for each dataset type. This can be solved in two different ways:</p>
<ul class="simple">
<li><p>Have a single Encoder class containing separate methods for encoding different dataset types.
During encoding, the dataset type is checked, and the corresponding methods are called.
An example of this is given in the SillyEncoder.</p></li>
<li><p>Have an abstract base Encoder class for the general encoding type, with subclasses for each dataset type.
The base Encoder contains all shared functionalities, and the subclasses contain dataset-specific functionalities,
such as code for ‘encoding an example’.
Note that in this case, the base Encoder implements the method <code class="code docutils literal notranslate"><span class="pre">build_object(dataset:</span> <span class="pre">Dataset,</span> <span class="pre">params)</span></code>,
that returns the correct dataset type-specific encoder subclass.
An example of this is <code class="xref py py-obj docutils literal notranslate"><span class="pre">OneHotEncoder</span></code>, which has subclasses <code class="xref py py-obj docutils literal notranslate"><span class="pre">OneHotSequenceEncoder</span></code>,
<code class="xref py py-obj docutils literal notranslate"><span class="pre">OneHotReceptorEncoder</span></code> and <code class="xref py py-obj docutils literal notranslate"><span class="pre">OneHotRepertoireEncoder</span></code></p></li>
</ul>
<p>When an encoding only makes sense for one possible dataset type, only one class needs to be created.
The <code class="code docutils literal notranslate"><span class="pre">build_object(dataset:</span> <span class="pre">Dataset,</span> <span class="pre">params)</span></code> method should raise a user-friendly error when an illegal dataset type is supplied.
An example of this can be found in <code class="xref py py-obj docutils literal notranslate"><span class="pre">SimilarToPositiveSequenceEncoder</span></code>.</p>
</section>
<section id="input-and-output-of-the-encode-method">
<h3>Input and output of the encode() method<a class="headerlink" href="#input-and-output-of-the-encode-method" title="Link to this heading">¶</a></h3>
<p>The encode() method is called by immuneML to encode a new dataset.
This method is called with two arguments: a dataset and params (an <code class="xref py py-obj docutils literal notranslate"><span class="pre">EncoderParams</span></code> object), which contains:</p>
<blockquote>
<div><p><strong>EncoderParams:</strong></p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">label_config</span></code>: a <code class="xref py py-obj docutils literal notranslate"><span class="pre">LabelConfiguration</span></code> object containing the labels that were specified for the analysis. Should be used as an input parameter for <code class="code docutils literal notranslate"><span class="pre">EncoderHelper.encode_dataset_labels()</span></code>.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">encode_labels</span></code>: boolean value which specifies whether labels must be used when encoding. Should be used as an input parameter for <code class="code docutils literal notranslate"><span class="pre">EncoderHelper.encode_dataset_labels()</span></code>.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">pool_size</span></code>: the number of parallel processes that the Encoder is allowed to use, for example when using parallelisation using the package <code class="code docutils literal notranslate"><span class="pre">pool</span></code>. This only needs to be used when implementing parallelisation.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">result_path</span></code>: this path can optionally be used to store intermediate files, if necessary. For most encoders, this is not necessary.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">learn_model</span></code>: a boolean value indicating whether the encoder is called during ‘training’ (learn_model=True) or ‘application’ (learn_model=False). Thus, this parameter can be used to prevent ‘leakage’ of information from the test to training set. This must be taken into account when performing operations over the whole dataset, such as normalising/scaling features (example: <code class="xref py py-obj docutils literal notranslate"><span class="pre">Word2VecEncoder</span></code>). For encoders where the encoding of a single example is not dependent on other examples, (e.g., <code class="xref py py-obj docutils literal notranslate"><span class="pre">OneHotEncoder</span></code>), this parameter can be ignored.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">model</span></code>: this parameter is used by e.g., <code class="xref py py-obj docutils literal notranslate"><span class="pre">KmerFrequencyEncoder</span></code> to pass its parameters to other classes. This parameter can usually be ignored.</p></li>
</ul>
</div></blockquote>
<p>The <code class="code docutils literal notranslate"><span class="pre">encode()</span></code> method should return a new dataset object, which is a copy of the original input dataset, but with an added <code class="code docutils literal notranslate"><span class="pre">encoded_data</span></code> attribute.
The <code class="code docutils literal notranslate"><span class="pre">encoded_data</span></code> attribute should contain an <code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodedData</span></code> object, which is created with the
following arguments:</p>
<blockquote>
<div><p><strong>EncodedData:</strong></p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">examples</span></code>: a design matrix where the rows represent Repertoires, Receptors or Sequences (‘examples’), and the columns the encoding-specific features. This is typically a numpy matrix, but may also be another matrix type (e.g., scipy sparse matrix, pytorch tensor, pandas dataframe, numpy.memmap if a lot of memory is required).</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">encoding</span></code>: a string denoting the encoder base class that was used.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">labels</span></code>: a dictionary of labels, where each label is a key, and the values are the label values across the examples (for example: {disease1: [positive, positive, negative]} if there are 3 repertoires). This parameter should be set only if <code class="code docutils literal notranslate"><span class="pre">EncoderParams.encode_labels</span></code> provided to encoder’s encode() function is set to True, otherwise it should be set to None. This can be created by calling utility function <code class="code docutils literal notranslate"><span class="pre">EncoderHelper.encode_dataset_labels()</span></code>.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">example_ids</span></code>: a list of identifiers for the examples (Repertoires, Receptors or Sequences). This can be retrieved using <code class="code docutils literal notranslate"><span class="pre">Dataset.get_example_ids()</span></code>.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">feature_names</span></code>: a list of feature names, i.e., the names given to the encoding-specific features. When included, list must be as long as the number of features.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">feature_annotations</span></code>: an optional pandas dataframe with additional information about the features. When included, number of rows in this dataframe must correspond to the number of features. This parameter is not typically used.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">info</span></code>: an optional dictionary that may be used to store any additional information that is relevant (for example paths to additional output files). This parameter is not typically used.</p></li>
</ul>
</div></blockquote>
<p>The <code class="code docutils literal notranslate"><span class="pre">examples</span></code> attribute of the <code class="code docutils literal notranslate"><span class="pre">EncodedData</span></code> objects will be directly passed to the ML models for training.
Other attributes are used for reports and interpretability.</p>
</section>
<section id="caching-intermediate-results">
<h3>Caching intermediate results<a class="headerlink" href="#caching-intermediate-results" title="Link to this heading">¶</a></h3>
<p>To prevent recomputing the same result a second time, immuneML uses caching.
Caching can be applied to methods which compute an (intermediate) result.
The result is stored to a file, and when the same method call is made, the previously
stored result is retrieved from the file and returned.</p>
<p>We recommend applying caching to methods which are computationally expensive and may be called
multiple times in the same way. For example, encoders are a good target for caching as they
may take long to compute and can be called multiple times on the same data when combined
with different ML methods. ML methods typically do not require caching, as you would
want to fit ML methods with different parameters or to differently encoded data.</p>
<p>Any method call in immuneML can be cached as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">CacheHandler</span><span class="o">.</span><span class="n">memo_by_params</span><span class="p">(</span><span class="n">params</span> <span class="o">=</span> <span class="n">cache_params</span><span class="p">,</span> <span class="n">fn</span> <span class="o">=</span> <span class="k">lambda</span><span class="p">:</span> <span class="n">my_method_for_caching</span><span class="p">(</span><span class="n">my_method_param1</span><span class="p">,</span> <span class="n">my_method_param2</span><span class="p">,</span> <span class="o">...</span><span class="p">))</span>
</pre></div>
</div>
<p>The <code class="code docutils literal notranslate"><span class="pre">CacheHandler.memo_by_params</span></code> method does the following:</p>
<ul class="simple">
<li><p>Using the caching parameters, a unique cache key (string) is created.</p></li>
<li><p>CacheHandler checks if there already exists a previously computed result that is associated with this key.</p></li>
<li><p>If the result exists, the result is returned without (re)computing the method.</p></li>
<li><p>If the result does not exist, the method is computed, its result is stored using the cache key, and the result is returned.</p></li>
</ul>
<p>The <code class="code docutils literal notranslate"><span class="pre">lambda</span></code> function call simply calls the method whose results will be cached, using any required parameters.
The <code class="code docutils literal notranslate"><span class="pre">cache_params</span></code> represent the unique, immutable parameters used to compute the cache key.
It should have the following properties:</p>
<ul class="simple">
<li><p>It must be a nested tuple containing <em>only</em> immutable items such as strings, booleans and integers.
It cannot contain mutable items like lists, dictionaries, sets and objects (they all need to be converted nested tuples of immutable items).</p></li>
<li><p>It should include <em>every</em> factor that can contribute to a difference in the results of the computed method.
For example, when caching the encode_data step, the following should be included:</p>
<ul>
<li><p>dataset descriptors (dataset id, example ids, dataset type),</p></li>
<li><p>encoding name,</p></li>
<li><p>labels,</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">EncoderParams.learn_model</span></code> if used,</p></li>
<li><p>all relevant input parameters to the encoder. Preferentially retrieved automatically (such as by <code class="code docutils literal notranslate"><span class="pre">vars(self)</span></code>),
as this ensures that if new parameters are added to the encoder, they are always added to the caching params.</p></li>
</ul>
</li>
</ul>
<p>For example, <code class="xref py py-obj docutils literal notranslate"><span class="pre">OneHotEncoder</span></code> computes its
caching parameters as follows:</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">_prepare_caching_params</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">params</span><span class="p">:</span> <span class="n">EncoderParams</span><span class="p">):</span>
    <span class="k">return</span> <span class="p">((</span><span class="s2">&quot;dataset_identifier&quot;</span><span class="p">,</span> <span class="n">dataset</span><span class="o">.</span><span class="n">identifier</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;example_identifiers&quot;</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">dataset</span><span class="o">.</span><span class="n">get_example_ids</span><span class="p">())),</span>
            <span class="p">(</span><span class="s2">&quot;dataset_type&quot;</span><span class="p">,</span> <span class="n">dataset</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;encoding&quot;</span><span class="p">,</span> <span class="n">OneHotEncoder</span><span class="o">.</span><span class="vm">__name__</span><span class="p">),</span>
            <span class="p">(</span><span class="s2">&quot;labels&quot;</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">label_config</span><span class="o">.</span><span class="n">get_labels_by_name</span><span class="p">())),</span>
            <span class="p">(</span><span class="s2">&quot;encoding_params&quot;</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">vars</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">())))</span>
</pre></div>
</div>
</div></blockquote>
<p>The construction of caching parameters must be done carefully, as caching bugs are extremely difficult
to discover. Rather add ‘too much’ information than too little.
A missing parameter will not lead to an error, but can result in silently copying over
results from previous method calls that do not match the current call.</p>
</section>
<section id="class-documentation-standards-for-encodings">
<h3>Class documentation standards for encodings<a class="headerlink" href="#class-documentation-standards-for-encodings" title="Link to this heading">¶</a></h3>
<p>Class documentation should be added as a docstring to all new Encoder, MLMethod, Report or Preprocessing classes.
The class docstrings are used to automatically generate the documentation web pages, using Sphinx <a class="reference external" href="https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html">reStructuredText</a>, and should adhere to a standard format:</p>
<ol class="arabic simple">
<li><p>A short, general description of the functionality</p></li>
<li><p>Optional extended description, including any references or specific cases that should be considered. For instance: if a class can only be used for a particular dataset type. Compatibility between Encoders, MLMethods and Reports should also be described.</p></li>
<li><p>For encoders, the appropriate dataset type(s). For example:</p></li>
</ol>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="err">*</span><span class="nv">*Dataset</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">type:**</span>

<span class="l l-Scalar l-Scalar-Plain">- SequenceDatasets</span>

<span class="l l-Scalar l-Scalar-Plain">- RepertoireDatasets</span>
</pre></div>
</div>
<ol class="arabic simple">
<li><p>A list of arguments, when applicable. This should follow the format below:</p></li>
</ol>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="err">*</span><span class="nv">*Specification</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">arguments:**</span>

<span class="l l-Scalar l-Scalar-Plain">- parameter_name (type)</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">a short description</span>

<span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">other_paramer_name (type)</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">a short description</span>
</pre></div>
</div>
<ol class="arabic simple">
<li><p>A YAML snippet, to show an example of how the new component should be called. Make sure to test your YAML snippet in an immuneML run to ensure it is specified correctly. The following formatting should be used to ensure the YAML snippet is rendered correctly:</p></li>
</ol>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="err">*</span><span class="nv">*YAML</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">specification:**</span>

<span class="l l-Scalar l-Scalar-Plain">.. code-block:</span><span class="p p-Indicator">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">yaml</span>

<span class="w">    </span><span class="l l-Scalar l-Scalar-Plain">definitions</span><span class="p p-Indicator">:</span>
<span class="w">        </span><span class="nt">yaml_keyword</span><span class="p">:</span><span class="w"> </span><span class="c1"># could be encodings/ml_methods/reports/etc...</span>
<span class="w">            </span><span class="nt">my_new_class</span><span class="p">:</span>
<span class="w">                </span><span class="nt">MyNewClass</span><span class="p">:</span>
<span class="w">                    </span><span class="nt">parameter_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">                    </span><span class="nt">other_paramer_name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
</pre></div>
</div>
<details class="summary-click-to-view-a-full-example-of-datasetencoder-class-documentation">
<summary>Click to view a full example of DatasetEncoder class documentation.</summary><div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">This</span> <span class="n">SillyEncoder</span> <span class="k">class</span> <span class="nc">is</span> <span class="n">a</span> <span class="n">placeholder</span> <span class="k">for</span> <span class="n">a</span> <span class="n">real</span> <span class="n">encoder</span><span class="o">.</span>
<span class="n">It</span> <span class="n">computes</span> <span class="n">a</span> <span class="nb">set</span> <span class="n">of</span> <span class="n">random</span> <span class="n">numbers</span> <span class="k">as</span> <span class="n">features</span> <span class="k">for</span> <span class="n">a</span> <span class="n">given</span> <span class="n">dataset</span><span class="o">.</span>

<span class="o">**</span><span class="n">Specification</span> <span class="n">arguments</span><span class="p">:</span><span class="o">**</span>

<span class="o">-</span> <span class="n">random_seed</span> <span class="p">(</span><span class="nb">int</span><span class="p">):</span> <span class="n">The</span> <span class="n">random</span> <span class="n">seed</span> <span class="k">for</span> <span class="n">generating</span> <span class="n">random</span> <span class="n">features</span><span class="o">.</span>

<span class="o">-</span> <span class="n">embedding_len</span> <span class="p">(</span><span class="nb">int</span><span class="p">):</span> <span class="n">The</span> <span class="n">number</span> <span class="n">of</span> <span class="n">random</span> <span class="n">features</span> <span class="n">to</span> <span class="n">generate</span> <span class="n">per</span> <span class="n">example</span><span class="o">.</span>


 <span class="o">**</span><span class="n">YAML</span> <span class="n">specification</span><span class="p">:</span><span class="o">**</span>

<span class="o">..</span> <span class="n">indent</span> <span class="k">with</span> <span class="n">spaces</span>
<span class="o">..</span> <span class="n">code</span><span class="o">-</span><span class="n">block</span><span class="p">::</span> <span class="n">yaml</span>

    <span class="n">definitions</span><span class="p">:</span>
        <span class="n">encodings</span><span class="p">:</span>
            <span class="n">my_silly_encoder</span><span class="p">:</span>
                <span class="n">Silly</span><span class="p">:</span>
                    <span class="n">random_seed</span><span class="p">:</span> <span class="mi">1</span>
                    <span class="n">embedding_len</span><span class="p">:</span> <span class="mi">5</span>
</pre></div>
</div>
</details></section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="how_to_add_new_ML_method.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">How to add a new machine learning method</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="install_for_development.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Set up immuneML for development</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2025, Milena Pavlovic, Lonneke Scheffer, Keshav Motwani, Victor Greiff, Geir Kjetil Sandve
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">How to add a new encoding</a><ul>
<li><a class="reference internal" href="#adding-an-example-encoder-to-the-immuneml-codebase">Adding an example encoder to the immuneML codebase</a><ul>
<li><a class="reference internal" href="#step-by-step-tutorial">Step-by-step tutorial</a></li>
<li><a class="reference internal" href="#test-running-the-new-encoding-with-a-yaml-specification">Test running the new encoding with a YAML specification</a></li>
<li><a class="reference internal" href="#adding-a-unit-test-for-a-datasetencoder">Adding a Unit test for a DatasetEncoder</a></li>
</ul>
</li>
<li><a class="reference internal" href="#implementing-a-new-encoder">Implementing a new encoder</a><ul>
<li><a class="reference internal" href="#encoders-for-different-dataset-types">Encoders for different dataset types</a></li>
<li><a class="reference internal" href="#input-and-output-of-the-encode-method">Input and output of the encode() method</a></li>
<li><a class="reference internal" href="#caching-intermediate-results">Caching intermediate results</a></li>
<li><a class="reference internal" href="#class-documentation-standards-for-encodings">Class documentation standards for encodings</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=f1196aca"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>