<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta content="summary" name="twitter:card" />
<meta content="&#64;immuneml" name="twitter:site" />
<meta content="immuneML quickstart: get started in Galaxy with YAML specification" name="twitter:title" />
<meta content="See tutorials on how to run an immuneML quickstart in Galaxy with a YAML specification." name="twitter:description" />
<meta content="https://docs.immuneml.uio.no/_images/receptor_classification_overview.png" name="twitter:image" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Quickstart: command-line interface with YAML" href="cli_yaml.html" /><link rel="prev" title="Quickstart: Galaxy with button-based tools" href="galaxy_simple.html" />
        <link rel="canonical" href="https://docs.immuneml.uio.no/quickstart/galaxy_yaml.html" />

    <link rel="shortcut icon" href="../_static/favicon.ico"/><!-- Generated with Sphinx 8.2.3 and Furo 2024.08.06 -->
        <title>Quickstart: Galaxy with YAML-based tools - immuneML 3.0.14 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=302659d7" />
    <link rel="stylesheet" type="text/css" href="../_static/css/immuneml.css" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">immuneML 3.0.14 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">immuneML 3.0.14 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 current has-children"><a class="reference internal" href="../quickstart.html">Quickstart</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Quickstart</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="galaxy_simple.html">Quickstart: Galaxy with button-based tools</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Quickstart: Galaxy with YAML-based tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli_yaml.html">Quickstart: command-line interface with YAML</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulation_quickstart.html">LIgO simulation quickstart</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../installation.html">Installing immuneML</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Installing immuneML</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation/install_with_package_manager.html">Install immuneML with a package manager</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/installation_docker.html">Setting up immuneML with Docker</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation/cloud.html">Running immuneML in the cloud</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../specification.html">YAML specification</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of YAML specification</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/how_to_specify_an_analysis_with_yaml.html">How to specify an analysis with YAML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/datasets.html">Dataset parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/encodings.html">Encoding parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/ml_methods.html">ML method parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/reports.html">Report parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/preprocessings.html">Preprocessing parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/simulation.html">Simulation parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../yaml_specs/instructions.html">Instruction parameters</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/analyze_dataset.html">Analyzing Your Own Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_import_the_data_to_immuneML.html">How to import data into immuneML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_generate_a_random_repertoire_dataset.html">How to generate a dataset with random sequences</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../tutorials/ligo_simulation_tutorials.html">Dataset simulation with LIgO</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of Dataset simulation with LIgO</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/ligo_simulation_yaml.html">YAML specification of the LigoSim instruction for introducing immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/how_to_simulate_co-occuring_signals.html">How to simulate co-occuring immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/how_to_simulate_paired_chain_data.html">Paired chain simulations in LIgO</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/simulation_with_custom_signal_functions.html">Simulation with custom signal functions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_train_and_assess_a_receptor_or_repertoire_classifier.html">How to train and assess a receptor or repertoire-level ML classifier</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_apply_to_new_data.html">How to apply previously trained ML models to a new dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/how_to_perform_exploratory_analysis.html">How to perform an exploratory data analysis</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../tutorials/motif_recovery.html">How to find motifs associated with disease or antigen binding state</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of How to find motifs associated with disease or antigen binding state</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/discover_motifs_precision_recall.html">Discovering positional motifs using precision and recall thresholds</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/discover_motifs_classifiers.html">Discovering motifs learned by classifiers</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/recovering_simulated_motifs.html">Recovering simulated immune signals</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/comparing_baseline_motifs.html">Comparing baseline motif frequencies in repertoires</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/clustering_tutorial.html">How to perform clustering analysis</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../galaxy.html">immuneML &amp; Galaxy</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of immuneML &amp; Galaxy</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_intro.html">Introduction to Galaxy</a></li>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_tools.html">immuneML Galaxy tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../galaxy/galaxy_immunologist_friendly.html">ML basics: Training classifiers with the simplified Galaxy interface</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../usecases.html">Use case examples</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of Use case examples</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../usecases/emerson_reproduction.html">Manuscript use case 1: Reproduction of a published study inside immuneML</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/extendability_use_case.html">Manuscript use case 2: Extending immuneML with a deep learning component for predicting antigen specificity of paired receptor data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/benchmarking_use_case.html">Manuscript use case 3: Benchmarking ML methods on ground-truth synthetic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immcantation_use_case.html">Integration use case: post-analysis of sequences with Immcantation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immunarch_use_case.html">Integration use case: post-analysis of sequences with immunarch</a></li>
<li class="toctree-l2"><a class="reference internal" href="../usecases/immunesim_use_case.html">Integration use case: Performing analysis on immuneSIM-generated repertoires</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../troubleshooting.html">Troubleshooting</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of Troubleshooting</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../developer_docs.html">Developer documentation</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of Developer documentation</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/info_new_developers.html">Information for new developers</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/install_for_development.html">Set up immuneML for development</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_encoding.html">How to add a new encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_ML_method.html">How to add a new machine learning method</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_report.html">How to add a new report</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/how_to_add_new_preprocessing.html">How to add a new preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/data_model.html">immuneML data model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../developer_docs/execution_flow.html">immuneML execution flow</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../_sources/quickstart/galaxy_yaml.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="quickstart-galaxy-with-yaml-based-tools">
<h1>Quickstart: Galaxy with YAML-based tools<a class="headerlink" href="#quickstart-galaxy-with-yaml-based-tools" title="Link to this heading">¶</a></h1>
<p>This tutorial will show you how to use immuneML for a simple machine learning analysis on an adaptive immune receptor repertoire (AIRR) dataset.
This example dataset consists of 100 synthetic immune repertoires (amino acid sequences generated by <a class="reference external" href="https://github.com/statbiophys/OLGA">OLGA</a>), each containing 1000 CDR3 sequences.
In half the repertoires, the subsequence ‘VLEQ’ has been implanted in 5% of the CDR3 sequences to simulate a disease signal.
Using immuneML, we will encode the data as 3-mer frequencies and train a logistic regression model to predict the disease status of each repertoire
(i.e., whether the repertoire contains the disease signal ‘VLEQ’), which is a binary classification problem.</p>
<section id="getting-started-through-galaxy">
<h2>Getting started through Galaxy<a class="headerlink" href="#getting-started-through-galaxy" title="Link to this heading">¶</a></h2>
<p>The Galaxy web interface is available at <a class="reference external" href="https://avant.immuneml.uiocloud.no/">https://avant.immuneml.uiocloud.no/</a>.
You may choose to register a user account or perform the analysis as an anonymous user.
If you are an anonymous user, the data will disappear once your browser session expires.</p>
<p>Different functionalities are available as Galaxy tools (left menu), and the analysis results appear in the Galaxy history (right list).
These history elements can be used as input for the next tool, creating a multi-step analysis workflow.</p>
<p>The purpose of the Quickstart tutorial is to quickly guide you through an analysis step by step, without assuming prior knowledge of Galaxy.
Additionally, you can check out our <a class="reference internal" href="../galaxy/galaxy_intro.html#introduction-to-galaxy"><span class="std std-ref">Introduction to Galaxy</span></a> for a more detailed explanation of the Galaxy interface.</p>
<section id="step-1-importing-the-dataset-to-a-galaxy-history">
<h3>Step 1: importing the dataset to a Galaxy history<a class="headerlink" href="#step-1-importing-the-dataset-to-a-galaxy-history" title="Link to this heading">¶</a></h3>
<p>Every immuneML analysis takes a dataset as input. For the Quickstart tutorial, an example dataset has been prepared and is
available through <a class="reference external" href="https://avant.immuneml.uiocloud.no/u/immuneml/h/quickstart-data">this Galaxy history</a> (under ‘Shared data’ &gt; ‘Histories’ &gt; ‘Quickstart data’).
Alternatively, the tutorial <span class="xref std std-ref">How to make an immuneML dataset in Galaxy</span> describes in detail
how to make an immuneML dataset using your own data.</p>
<p>Individual files can be inspected by clicking the eyeball icons.
To import the complete history, click the + icon in the right upper corner.</p>
<a class="reference internal image-reference" href="../_images/import_galaxy_history.png"><img alt="import button Galaxy" src="../_images/import_galaxy_history.png" style="width: 250px;" />
</a>
<p>This Quickstart dataset Galaxy history contains the following items:</p>
<ul class="simple">
<li><p>100 repertoire .tsv files in AIRR format. For details about the AIRR format, see the <a class="reference external" href="https://docs.airr-community.org/en/stable/datarep/format.html">AIRR documentation</a></p></li>
<li><p>A Collection of repertoires. This history element collects the 100 above-mentioned repertoire files in a Galaxy collection.
This Galaxy collection makes it easier to select the repertoires as an input to Galaxy tools (instead of selecting all 100 files manually, you can select the collection).
To read how to make your own Galaxy collection, see <a class="reference internal" href="../galaxy/galaxy_intro.html#creating-a-galaxy-collection-of-files"><span class="std std-ref">Creating a Galaxy collection of files</span></a>.</p></li>
<li><p>A metadata.csv file. The metadata file describes which of the 100 repertoires are diseased and healthy, under the
column named ‘signal_disease’ which contains the values True and False.
For details about the metadata file, see <a class="reference internal" href="../tutorials/how_to_import_the_data_to_immuneML.html#what-should-the-metadata-file-look-like"><span class="std std-ref">What should the metadata file look like?</span></a></p></li>
</ul>
</section>
<section id="step-2-creating-an-immuneml-galaxy-dataset">
<h3>Step 2: creating an immuneML Galaxy dataset<a class="headerlink" href="#step-2-creating-an-immuneml-galaxy-dataset" title="Link to this heading">¶</a></h3>
<p>Use the <a class="reference external" href="https://avant.immuneml.uiocloud.no/root?tool_id=immuneml_dataset">Create Dataset with Reports</a> Galaxy tool (under ‘immuneML tools’) to import the dataset
and create an <em>immuneML dataset</em> history item, which can subsequently be used as input for other Galaxy tools.</p>
<p>Select ‘Simplified interface’, then ‘repertoire dataset’ as dataset type and ‘AIRR’ data format, and select the metadata.csv file as metadata file.
Under the ‘Data files’ field, all repertoire files must be selected.
This can either be done by manually selecting all 100 repertoire files, or by selecting the Collection of repertoires.
For convenience, we will use the latter option.</p>
<p>By default, the ‘Data files’ menu will show all txt-like files (such as repertoire files) that are present in the history.
By clicking the ‘Dataset collections’ button (folder icon), the menu will instead only show the collections in the history.
Select the collection of repertoires so it becomes highlighted.</p>
<a class="reference internal image-reference" href="../_images/create_dataset_from_collection.png"><img alt="create dataset from collection" src="../_images/create_dataset_from_collection.png" style="width: 500px;" />
</a>
<p>Finally, click ‘execute’.</p>
<p>Three new items will appear in the Galaxy history. If everything went correctly, the items would appear green. If the items
are red, an error occurred (check if you correctly selected the collection of repertoires).</p>
<p>In ‘Summary: dataset generation’ you can find details about the newly
generated dataset, including the name of the dataset, the dataset type and size, and a download link.</p>
<p>The next item, ‘create_dataset.yaml’ the YAML specification file that was used by immuneML to create the dataset.
This YAML specification could be used when running the Create dataset tool with the ‘Advanced’ interface.</p>
<p>Finally, ‘immuneML dataset’ is a new Galaxy collection containing the immuneML dataset in the ImmuneML format.
The ImmuneML format is not human-readable, but it ensures that you can quickly import the dataset into various Galaxy tools
without having to repeatedly specify the import parameters.</p>
</section>
<section id="step-3-writing-the-yaml-specification">
<h3>Step 3: writing the YAML specification<a class="headerlink" href="#step-3-writing-the-yaml-specification" title="Link to this heading">¶</a></h3>
<p>Any immuneML analysis is described by a YAML specification file.
This file contains nested key-value pairs. Mandatory keywords with a specific meaning are <code class="code docutils literal notranslate"><span class="pre">styled</span> <span class="pre">like</span> <span class="pre">this</span></code>
in the text. Note that correct whitespace (not tab) indentation of the yaml file is important.</p>
<p>In this tutorial, we will only cover the essential elements of the YAML specification.
For a more complete introduction, see <a class="reference internal" href="../yaml_specs/how_to_specify_an_analysis_with_yaml.html#how-to-specify-an-analysis-with-yaml"><span class="std std-ref">How to specify an analysis with YAML</span></a>.</p>
<p>The YAML specification consists of:</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">definitions</span></code> describing the analysis components.</p>
<ul>
<li><p><code class="code docutils literal notranslate"><span class="pre">datasets</span></code>: our data is in <a class="reference internal" href="../yaml_specs/datasets.html#airr"><span class="std std-ref">AIRR</span></a> format, we need to provide the name of the metadata file.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">encodings</span></code>: the data will be represented through a <a class="reference internal" href="../yaml_specs/encodings.html#kmerfrequency"><span class="std std-ref">k-mer frequency encoding</span></a>.
This means each repertoire is represented based on the frequency of subsequences of length k.
For example, the sequence CSVQYF contains the 3-mers CSV, SVQ, VQY and QYF.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">ml_methods</span></code>: we will use <a class="reference internal" href="../yaml_specs/ml_methods.html#logisticregression"><span class="std std-ref">logistic regression</span></a> to classify
the encoded immune repertoires.</p></li>
<li><p>Optionally, <code class="code docutils literal notranslate"><span class="pre">reports</span></code>: we will plot the <a class="reference internal" href="../yaml_specs/reports.html#coefficients"><span class="std std-ref">coefficients</span></a> of the trained
logistic regression model, to get more insight into what the model has learned.</p></li>
</ul>
</li>
<li><p><code class="code docutils literal notranslate"><span class="pre">instructions</span></code> describing the type of analysis.</p>
<ul>
<li><p>The <a class="reference internal" href="../yaml_specs/instructions.html#trainmlmodel"><span class="std std-ref">TrainMLModel</span></a> instruction is used to train one or more ‘ML settings’ (combinations of encodings and ML methods),
and optimize the hyperparameters using nested cross-validation. We can set the parameters for the outer ‘assessment’ and
inner ‘selection’ cross-validation loops.</p></li>
</ul>
</li>
</ul>
<p>The complete YAML specification for this analysis looks like this and can be downloaded here: <a class="reference download internal" download="" href="../_downloads/21d94bd5588c53f09c040bee89c62b62/quickstart.yaml"><code class="xref download docutils literal notranslate"><span class="pre">quickstart.yaml</span></code></a>.</p>
<details class="summary-quickstart-yaml">
<summary>quickstart.yaml</summary><div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">definitions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">datasets</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_dataset</span><span class="p">:</span><span class="w"> </span><span class="c1"># user-defined dataset name</span>
<span class="w">      </span><span class="nt">format</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">AIRR</span>
<span class="w">      </span><span class="nt">params</span><span class="p">:</span>
<span class="w">        </span><span class="nt">dataset_file</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">dataset.yaml</span><span class="w"> </span><span class="c1"># &#39;dataset&#39; is the default name given by the Create dataset tool</span>

<span class="w">  </span><span class="nt">encodings</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_kmer_frequency</span><span class="p">:</span><span class="w"> </span><span class="c1"># user-defined encoding name</span>
<span class="w">      </span><span class="nt">KmerFrequency</span><span class="p">:</span><span class="w">   </span><span class="c1"># encoding type</span>
<span class="w">        </span><span class="nt">k</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span><span class="w">           </span><span class="c1"># encoding parameters</span>

<span class="w">  </span><span class="nt">ml_methods</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_logistic_regression</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">LogisticRegression</span><span class="w"> </span><span class="c1"># user-defined ML model name: ML model type (no user-specified parameters)</span>

<span class="w">  </span><span class="nt">reports</span><span class="p">:</span>
<span class="w">    </span><span class="nt">my_benchmark</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">MLSettingsPerformance</span><span class="w"> </span><span class="c1"># user-defined report name: report type (no user-specified parameters)</span>
<span class="w">    </span><span class="nt">my_coefficients</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Coefficients</span>

<span class="nt">instructions</span><span class="p">:</span>
<span class="w">  </span><span class="nt">my_training_instruction</span><span class="p">:</span><span class="w"> </span><span class="c1"># user-defined instruction name</span>
<span class="w">    </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">TrainMLModel</span>

<span class="w">    </span><span class="nt">dataset</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_dataset</span><span class="w"> </span><span class="c1"># use the same dataset name as in definitions</span>
<span class="w">    </span><span class="nt">labels</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">signal_disease</span><span class="w">    </span><span class="c1"># use a label available in the metadata.csv file</span>

<span class="w">    </span><span class="nt">settings</span><span class="p">:</span><span class="w"> </span><span class="c1"># which combinations of ML settings to run</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">encoding</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_kmer_frequency</span>
<span class="w">      </span><span class="nt">ml_method</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_logistic_regression</span>

<span class="w">    </span><span class="nt">assessment</span><span class="p">:</span><span class="w"> </span><span class="c1"># parameters in the assessment (outer) cross-validation loop</span>
<span class="w">      </span><span class="nt">reports</span><span class="p">:</span><span class="w">  </span><span class="c1"># plot the coefficients for the trained model</span>
<span class="w">        </span><span class="nt">models</span><span class="p">:</span>
<span class="w">        </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_coefficients</span>
<span class="w">      </span><span class="nt">split_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">random</span><span class="w">   </span><span class="c1"># how to split the data - here: split randomly</span>
<span class="w">      </span><span class="nt">split_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w">           </span><span class="c1"># how many times (here once - just to train and test)</span>
<span class="w">      </span><span class="nt">training_percentage</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.7</span><span class="w"> </span><span class="c1"># use 70% of the data for training</span>

<span class="w">    </span><span class="nt">selection</span><span class="p">:</span><span class="w"> </span><span class="c1"># parameters in the selection (inner) cross-validation loop</span>
<span class="w">      </span><span class="nt">split_strategy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">random</span>
<span class="w">      </span><span class="nt">split_count</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">      </span><span class="nt">training_percentage</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span><span class="w"> </span><span class="c1"># use all data for training</span>

<span class="w">    </span><span class="nt">reports</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">my_benchmark</span>

<span class="w">    </span><span class="nt">optimization_metric</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">balanced_accuracy</span><span class="w"> </span><span class="c1"># the metric to optimize during nested cross-validation when comparing multiple models</span>
<span class="w">    </span><span class="nt">metrics</span><span class="p">:</span><span class="w"> </span><span class="c1"># other metrics to compute for reference</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">auc</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">precision</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">recall</span>

<span class="w">    </span><span class="nt">number_of_processes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">8</span><span class="w">     </span><span class="c1"># processes for parallelization</span>
</pre></div>
</div>
</details><p>The YAML specification can either be saved to a local file and uploaded to Galaxy, or pasted directly as a new entry.
This can be done by clicking ‘Upload Data’ in the left-hand menu, and choosing either ‘Choose local files’ or ‘Paste/Fetch data’.
The file will appear as a new history element</p>
<a class="reference internal image-reference" href="../_images/galaxy_upload_data.png"><img alt="upload data Galaxy" src="../_images/galaxy_upload_data.png" style="width: 250px;" />
</a>
</section>
<section id="step-4-running-the-analysis">
<h3>Step 4: running the analysis<a class="headerlink" href="#step-4-running-the-analysis" title="Link to this heading">¶</a></h3>
<p>Next, the <a class="reference external" href="https://avant.immuneml.uiocloud.no/root?tool_id=immuneml_train_ml_model">Train ML Classifiers</a> Galaxy tool (under ‘immuneML tools’) should be used.
Select the YAML specification and previously created dataset from the history, and no additional files.</p>
<a class="reference internal image-reference" href="../_images/galaxy_train_ml_model.png"><img alt="train ML model tool" src="../_images/galaxy_train_ml_model.png" style="width: 500px;" />
</a>
</section>
<section id="step-5-understanding-the-results">
<h3>Step 5: understanding the results<a class="headerlink" href="#step-5-understanding-the-results" title="Link to this heading">¶</a></h3>
<p>After running the Galaxy tool for training ML models, three new items will appear in the Galaxy history.</p>
<a class="reference internal image-reference" href="../_images/galaxy_train_ml_model_results.png"><img alt="history elements" src="../_images/galaxy_train_ml_model_results.png" style="width: 250px;" />
</a>
<p>The history item ‘Summary: ML model training’ contains HTML pages displaying a summary of the analysis.
On the first page, you will find a table which shows for each of the splits which of the ML settings (combination of ML model and encoding)
performed best in the inner cross-validation loop, and the performance score of that ML model on the test set.
Note that in this tutorial, only 1 data split was used.
Furthermore, you can click ‘see details’ in the table to find a more detailed explanation about how well each model
performed in the inner loop of cross validation (on the validation sets).
By clicking ‘see reports’ you can find a report that plots the 25 top coefficients of the trained logistic regression model.
Notice how the coefficients with the highest values are associated with the k-mers ‘VLE’ and ‘LEQ’, which overlap with
the implanted disease signal ‘VLEQ’, meaning the ML model learned the correct signal.</p>
<a class="reference internal image-reference" href="../_images/coefficients_logistic_regression.png"><img alt="coefficients report" src="../_images/coefficients_logistic_regression.png" style="width: 600px;" />
</a>
<p>The next item, “Archive: ML model training” contains a downloadable archive of the complete immuneML ouput (including
the files available through the other history elements).</p>
<p>Finally, history element “optimal_ml_settings.zip” is a .zip file containing the configuration of the optimal ML settings,
including settings for the encoding and machine learning method. This .zip file can be used to apply a previously
trained ML model to a new dataset, which can currently only be done <a href="#id1"><span class="problematic" id="id2">:ref:`How to apply previously trained ML models to a new dataset &lt;using the command-line interface&gt;`_</span></a>.</p>
</section>
</section>
<section id="what-s-next">
<h2>What’s next?<a class="headerlink" href="#what-s-next" title="Link to this heading">¶</a></h2>
<p>If you haven’t done it already, it is highly recommended to follow the <a class="reference internal" href="../galaxy/galaxy_intro.html#introduction-to-galaxy"><span class="std std-ref">Introduction to Galaxy</span></a> and <a class="reference internal" href="../yaml_specs/how_to_specify_an_analysis_with_yaml.html#how-to-specify-an-analysis-with-yaml"><span class="std std-ref">How to specify an analysis with YAML</span></a>.
If you want to try running immuneML on your own dataset, be sure to check out <span class="xref std std-ref">How to make an immuneML dataset in Galaxy</span>.
Other tutorials for how to use each of the immuneML Galaxy tools can be found under <a class="reference internal" href="../galaxy.html#immuneml-galaxy"><span class="std std-ref">immuneML &amp; Galaxy</span></a>.</p>
<p>The Galaxy interface is intended to make it easy for users to try out immuneML quickly, but for large-scale analyses,
please <a href="#id3"><span class="problematic" id="id4">:ref:`install immuneML &lt;Installing immuneML&gt;`_</span></a> locally or on a private server. YAML specifications are defined equivalently in
Galaxy and the command-line interface.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="cli_yaml.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Quickstart: command-line interface with YAML</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="galaxy_simple.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Quickstart: Galaxy with button-based tools</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2025, Milena Pavlovic, Lonneke Scheffer, Keshav Motwani, Victor Greiff, Geir Kjetil Sandve
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Quickstart: Galaxy with YAML-based tools</a><ul>
<li><a class="reference internal" href="#getting-started-through-galaxy">Getting started through Galaxy</a><ul>
<li><a class="reference internal" href="#step-1-importing-the-dataset-to-a-galaxy-history">Step 1: importing the dataset to a Galaxy history</a></li>
<li><a class="reference internal" href="#step-2-creating-an-immuneml-galaxy-dataset">Step 2: creating an immuneML Galaxy dataset</a></li>
<li><a class="reference internal" href="#step-3-writing-the-yaml-specification">Step 3: writing the YAML specification</a></li>
<li><a class="reference internal" href="#step-4-running-the-analysis">Step 4: running the analysis</a></li>
<li><a class="reference internal" href="#step-5-understanding-the-results">Step 5: understanding the results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#what-s-next">What’s next?</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=2efb4362"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>